{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "ClickTagDemuliplex20191009.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/GPCTP_2019/blob/master/Colab%20Notebooks/96SampleMuliplexExperiment.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XJT3cizx7aSA",
        "colab_type": "text"
      },
      "source": [
        "This Google Colab notebook contains code to reproduce analysis of a 96-sample multiplexed single-cell RNA-seq experiment performed in Gehring et. al 2019. \n",
        "\n",
        "In this experiment, mouse neural stem cells were treated with a combinatorial titration series of various growth factors, producing a 4x4x3x2 matrix of experimental conditions. Cells were collected and chemically cross-linked with poly-adenylated DNA oligos specific to each sample. After pooling and library preparation (10x Genomics v2 chemistry), these \"ClickTags\" are used to demultiplex cells followed by exploratory data analysis. \n",
        "\n",
        "This notebook includes code to downloaded and install all required software, and all data and metadata are downloaded as part of the workflow. The raw data are in the form of demultiplexed fastq files from Illumina DNA sequencing runs, and the metadata are standard files for processing single-cell RNA-seq such as a cell barcode whitelist, a ClickTag whitelist, and genome annotation files.\n",
        "\n",
        "Rather than requiring interested scientists to piece together our disparate data, analysis, and methods resources, we believe that open, fully reproducible data analysis like the one presented here will facilitate further development of analytical tools by the bioinformatics community as well as encourage experimentalists to carefully plan, execute, and analyze their experiments."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jNJjxyOx7beY",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Download software and required metadata\n",
        "---\n",
        "\n",
        "kallisto, bustools\n",
        "\n",
        "10x v2 whitelist\n",
        "\n",
        "mus musculus kallisto index\n",
        "\n",
        "mus musculus t2g (transcripts_to_genes) file\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BjpQA_0m0W9a",
        "colab_type": "code",
        "outputId": "fcdac6e3-508b-4858-bcca-624aadc5768b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "!date"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Fri Oct 18 19:01:40 UTC 2019\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eFhrZVnqzet2",
        "colab_type": "code",
        "outputId": "9d5ceaa7-7fae-468a-af5f-175d9f6dd7d7",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "!ls"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "sample_data\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ns5YpKWtSkqV",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!rm -r ./sample_data"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "k-hrrh600kjA",
        "colab_type": "code",
        "outputId": "1cdea9bf-a54b-4e7f-efbb-7d9929285efa",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 581
        }
      },
      "source": [
        "#DOWNLOAD and INSTALL kallisto and bustools\n",
        "\n",
        "!wget https://github.com/pachterlab/kallisto/releases/download/v0.46.0/kallisto_linux-v0.46.0.tar.gz\n",
        "!tar -xf kallisto_linux-v0.46.0.tar.gz\n",
        "!cp kallisto/kallisto /usr/local/bin/\n",
        "\n",
        "!wget https://github.com/BUStools/bustools/releases/download/v0.39.3/bustools_linux-v0.39.3.tar.gz\n",
        "!tar -xf bustools_linux-v0.39.3.tar.gz\n",
        "!cp bustools/bustools /usr/local/bin/"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-10-18 19:01:53--  https://github.com/pachterlab/kallisto/releases/download/v0.46.0/kallisto_linux-v0.46.0.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.118.4\n",
            "Connecting to github.com (github.com)|140.82.118.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/26562905/8cbbf280-8ca9-11e9-8c32-bec32f378e41?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190153Z&X-Amz-Expires=300&X-Amz-Signature=836cbc6015702e5b02f17db4a328ad53c420793f42a63e0395e4a4ca6c4ff9d9&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dkallisto_linux-v0.46.0.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-10-18 19:01:53--  https://github-production-release-asset-2e65be.s3.amazonaws.com/26562905/8cbbf280-8ca9-11e9-8c32-bec32f378e41?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190153Z&X-Amz-Expires=300&X-Amz-Signature=836cbc6015702e5b02f17db4a328ad53c420793f42a63e0395e4a4ca6c4ff9d9&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dkallisto_linux-v0.46.0.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.144.251\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.144.251|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 7260524 (6.9M) [application/octet-stream]\n",
            "Saving to: ‘kallisto_linux-v0.46.0.tar.gz’\n",
            "\n",
            "kallisto_linux-v0.4 100%[===================>]   6.92M  10.2MB/s    in 0.7s    \n",
            "\n",
            "2019-10-18 19:01:54 (10.2 MB/s) - ‘kallisto_linux-v0.46.0.tar.gz’ saved [7260524/7260524]\n",
            "\n",
            "--2019-10-18 19:02:00--  https://github.com/BUStools/bustools/releases/download/v0.39.3/bustools_linux-v0.39.3.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.118.4\n",
            "Connecting to github.com (github.com)|140.82.118.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/157250571/bf828d00-b3b4-11e9-801d-3356b95223ea?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190200Z&X-Amz-Expires=300&X-Amz-Signature=5e4178cd345441db829d86d3a872caa43cf659cdd5e5aa03a0b997491cbd70d0&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dbustools_linux-v0.39.3.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-10-18 19:02:00--  https://github-production-release-asset-2e65be.s3.amazonaws.com/157250571/bf828d00-b3b4-11e9-801d-3356b95223ea?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190200Z&X-Amz-Expires=300&X-Amz-Signature=5e4178cd345441db829d86d3a872caa43cf659cdd5e5aa03a0b997491cbd70d0&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dbustools_linux-v0.39.3.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.81.192\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.81.192|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 403317 (394K) [application/octet-stream]\n",
            "Saving to: ‘bustools_linux-v0.39.3.tar.gz’\n",
            "\n",
            "bustools_linux-v0.3 100%[===================>] 393.86K  1.20MB/s    in 0.3s    \n",
            "\n",
            "2019-10-18 19:02:01 (1.20 MB/s) - ‘bustools_linux-v0.39.3.tar.gz’ saved [403317/403317]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "O4VmCA1j0n6g",
        "colab_type": "code",
        "outputId": "b9753016-1b82-4b64-fba6-49470d7dacaa",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 629
        }
      },
      "source": [
        "#TEST kallisto and bustools\n",
        "!kallisto\n",
        "!bustools"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "kallisto 0.46.0\n",
            "\n",
            "Usage: kallisto <CMD> [arguments] ..\n",
            "\n",
            "Where <CMD> can be one of:\n",
            "\n",
            "    index         Builds a kallisto index \n",
            "    quant         Runs the quantification algorithm \n",
            "    bus           Generate BUS files for single-cell data \n",
            "    pseudo        Runs the pseudoalignment step \n",
            "    merge         Merges several batch runs \n",
            "    h5dump        Converts HDF5-formatted results to plaintext\n",
            "    inspect       Inspects and gives information about an index\n",
            "    version       Prints version information\n",
            "    cite          Prints citation information\n",
            "\n",
            "Running kallisto <CMD> without arguments prints usage information for <CMD>\n",
            "\n",
            "bustools 0.39.3\n",
            "\n",
            "Usage: bustools <CMD> [arguments] ..\n",
            "\n",
            "Where <CMD> can be one of: \n",
            "\n",
            "capture         Capture records from a BUS file\n",
            "correct         Error correct a BUS file\n",
            "count           Generate count matrices from a BUS file\n",
            "inspect         Produce a report summarizing a BUS file\n",
            "linker          Remove section of barcodes in BUS files\n",
            "project         Project a BUS file to gene sets\n",
            "sort            Sort a BUS file by barcodes and UMIs\n",
            "text            Convert a binary BUS file to a tab-delimited text file\n",
            "whitelist       Generate a whitelist from a BUS file\n",
            "\n",
            "Running bustools <CMD> without arguments prints usage information for <CMD>\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "K4resks4QGS9",
        "colab_type": "code",
        "outputId": "abe7091f-5c94-4c4b-d516-33019637cbb2",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 309
        }
      },
      "source": [
        "# DOWLOAD kallisto index\n",
        "!wget https://github.com/BUStools/getting_started/releases/download/getting_started/Mus_musculus.GRCm38.cdna.all.idx.gz\n",
        "!gunzip Mus_musculus.GRCm38.cdna.all.idx.gz"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-10-18 19:02:12--  https://github.com/BUStools/getting_started/releases/download/getting_started/Mus_musculus.GRCm38.cdna.all.idx.gz\n",
            "Resolving github.com (github.com)... 140.82.118.4\n",
            "Connecting to github.com (github.com)|140.82.118.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/9ba97600-8aec-11e9-9097-5b2f52d624f1?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190212Z&X-Amz-Expires=300&X-Amz-Signature=cd19c74d1416177239c694e42c29e46b76a1becf1e213377d490c53b466894b9&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3DMus_musculus.GRCm38.cdna.all.idx.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-10-18 19:02:12--  https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/9ba97600-8aec-11e9-9097-5b2f52d624f1?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190212Z&X-Amz-Expires=300&X-Amz-Signature=cd19c74d1416177239c694e42c29e46b76a1becf1e213377d490c53b466894b9&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3DMus_musculus.GRCm38.cdna.all.idx.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.101.27\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.101.27|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1687475406 (1.6G) [application/octet-stream]\n",
            "Saving to: ‘Mus_musculus.GRCm38.cdna.all.idx.gz’\n",
            "\n",
            "Mus_musculus.GRCm38 100%[===================>]   1.57G  11.2MB/s    in 51s     \n",
            "\n",
            "2019-10-18 19:03:03 (31.8 MB/s) - ‘Mus_musculus.GRCm38.cdna.all.idx.gz’ saved [1687475406/1687475406]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bCdT6Mhb0pwW",
        "colab_type": "code",
        "outputId": "5102f845-6091-4fb3-e0ca-c342fd3e7f74",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 309
        }
      },
      "source": [
        "#DOWNLOAD the whitelist and tx to gene file\n",
        "!wget https://github.com/BUStools/getting_started/releases/download/getting_started/10xv2_whitelist.txt"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-10-18 19:03:33--  https://github.com/BUStools/getting_started/releases/download/getting_started/10xv2_whitelist.txt\n",
            "Resolving github.com (github.com)... 140.82.118.3\n",
            "Connecting to github.com (github.com)|140.82.118.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/9ba97600-8aec-11e9-8c20-ab4bb3679829?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190333Z&X-Amz-Expires=300&X-Amz-Signature=4319797e4ea8097d12a772a429b9fac006bb028484e836fb4678bb4c9cbdf4ef&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3D10xv2_whitelist.txt&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-10-18 19:03:33--  https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/9ba97600-8aec-11e9-8c20-ab4bb3679829?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190333Z&X-Amz-Expires=300&X-Amz-Signature=4319797e4ea8097d12a772a429b9fac006bb028484e836fb4678bb4c9cbdf4ef&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3D10xv2_whitelist.txt&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.84.56\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.84.56|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 12533760 (12M) [application/octet-stream]\n",
            "Saving to: ‘10xv2_whitelist.txt’\n",
            "\n",
            "10xv2_whitelist.txt 100%[===================>]  11.95M  14.3MB/s    in 0.8s    \n",
            "\n",
            "2019-10-18 19:03:34 (14.3 MB/s) - ‘10xv2_whitelist.txt’ saved [12533760/12533760]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IVWZYBBnDkKa",
        "colab_type": "code",
        "outputId": "d0bc913b-14e8-4298-aa02-bbd328d8ae3c",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 309
        }
      },
      "source": [
        "!wget https://github.com/BUStools/getting_started/releases/download/getting_started/transcripts_to_genes.txt"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-10-18 19:03:35--  https://github.com/BUStools/getting_started/releases/download/getting_started/transcripts_to_genes.txt\n",
            "Resolving github.com (github.com)... 140.82.118.3\n",
            "Connecting to github.com (github.com)|140.82.118.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/2b9bef80-8aee-11e9-8f9b-954abf2a7ddb?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190336Z&X-Amz-Expires=300&X-Amz-Signature=71ca3937f41862200a57159d1af496e4cf6a231e0c3a37b035306e0f0d35ecbe&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dtranscripts_to_genes.txt&response-content-type=application%2Foctet-stream [following]\n",
            "--2019-10-18 19:03:36--  https://github-production-release-asset-2e65be.s3.amazonaws.com/191064839/2b9bef80-8aee-11e9-8f9b-954abf2a7ddb?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20191018%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191018T190336Z&X-Amz-Expires=300&X-Amz-Signature=71ca3937f41862200a57159d1af496e4cf6a231e0c3a37b035306e0f0d35ecbe&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dtranscripts_to_genes.txt&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.84.56\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.84.56|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 7063053 (6.7M) [application/octet-stream]\n",
            "Saving to: ‘transcripts_to_genes.txt’\n",
            "\n",
            "transcripts_to_gene 100%[===================>]   6.74M  10.1MB/s    in 0.7s    \n",
            "\n",
            "2019-10-18 19:03:37 (10.1 MB/s) - ‘transcripts_to_genes.txt’ saved [7063053/7063053]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OjbYuxY2kKIr",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Structure of the Raw Data\n",
        "---\n",
        "After ClickTag labeling, cells from all 96 samples were pooled, counted, and split across two lanes of the 10x Genomics v2 3' Gene Expression kit. After emulsification, each lane was split into a small and large fraction for sequencing on MiSeq and HiSeq, respectively. This procedure resulted in four sets of fastq files for ClickTag data (2 from MiSeq, 2 from HiSeq), and four corresponding fastq files for the cDNA reads. ClickTag data, which are used for sample assignment alone, do not need to be normalized prior to combining the datasets. The cDNA data, however, should be normalized to an equivalent number of reads per cell prior to combining cells from different libraries (See cDNA Cell Filtering and Normalization section below). "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GRMD-6-B7xM9",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Download the ClickTag data\n",
        "---\n",
        "fastq files\n",
        "\n",
        "ClickTag whitelist\n",
        "\n",
        "featuremap.py program"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "umQaVxblSjUD",
        "colab_type": "code",
        "outputId": "23de4eac-df62-4565-c909-b7124b8d3273",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 425
        }
      },
      "source": [
        "#DOWNLAOD the ClickTag fastqs\n",
        "\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/bac30204-8e17-4065-9ecd-f61d91b98ba5/ -o 96S_BIG2_tags_S1_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/7bf4afa7-b93b-4bbd-b184-5fe1dc109192/ -o 96S_BIG2_tags_S1_L001_R2_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/081b0d12-5bc9-4fff-9f61-fea051c6f8d0/ -o 96S_SMALL2_tags_S2_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/14c23f7b-9eed-4914-b795-0cfaadcedddd/ -o 96S_SMALL2_tags_S2_L001_R2_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/79d25783-7551-448f-b9cc-671db8278f21/ -o 96S_TagsBig1_S1_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/8e3d8fe5-1689-495d-9f98-dff15518d5ae/ -o 96S_TagsBig1_S1_L001_R2_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/f4f685dc-37f0-4f0e-ad78-1777cee12067/ -o 96S_tags_S1_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/d64f476e-7d94-4a35-b8ae-06d7cb9d050c/ -o 96S_tags_S1_L001_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  486M  100  486M    0     0  19.6M      0  0:00:24  0:00:24 --:--:-- 21.2M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 1488M  100 1488M    0     0  20.8M      0  0:01:11  0:01:11 --:--:-- 21.0M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 44.7M  100 44.7M    0     0  13.4M      0  0:00:03  0:00:03 --:--:-- 13.4M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  137M  100  137M    0     0  17.9M      0  0:00:07  0:00:07 --:--:-- 21.2M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  212M  100  212M    0     0  18.9M      0  0:00:11  0:00:11 --:--:-- 21.2M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  712M  100  712M    0     0  15.5M      0  0:00:45  0:00:45 --:--:-- 13.0M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 61.4M  100 61.4M    0     0  12.5M      0  0:00:04  0:00:04 --:--:-- 13.0M\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  172M  100  172M    0     0  8043k      0  0:00:21  0:00:21 --:--:-- 6794k\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vIWZ-F6ShsHz",
        "colab_type": "text"
      },
      "source": [
        "The ClickTag whitelist is simply a csv file containing the names and barcode sequences of each ClickTag used for sample multiplexing as well as an optional header."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "svaxiGexZdS6",
        "colab_type": "code",
        "outputId": "c95a57fe-f8c8-437c-80ec-677f13aea971",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 68
        }
      },
      "source": [
        "#DOWNLOAD the ClickTag Whitelist\n",
        "\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/febb5c97-3827-4219-940c-8b87cae3b77b/ -o ClickTagWhitelist.csv"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r100   498  100   498    0     0   1699      0 --:--:-- --:--:-- --:--:--  1699\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "73eg0Y7AayEy",
        "colab_type": "code",
        "outputId": "163c51e9-ef29-4858-eb9f-152cfc1fcf02",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 204
        }
      },
      "source": [
        "#DOWNLOAD featuremap.py\n",
        "\n",
        "!wget https://raw.githubusercontent.com/pachterlab/kite/master/featuremap/featuremap.py"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-10-18 19:07:06--  https://raw.githubusercontent.com/pachterlab/kite/master/featuremap/featuremap.py\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4673 (4.6K) [text/plain]\n",
            "Saving to: ‘featuremap.py’\n",
            "\n",
            "\rfeaturemap.py         0%[                    ]       0  --.-KB/s               \rfeaturemap.py       100%[===================>]   4.56K  --.-KB/s    in 0s      \n",
            "\n",
            "2019-10-18 19:07:06 (65.4 MB/s) - ‘featuremap.py’ saved [4673/4673]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rcxzZRir8uki",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Run featuremap.py followed by kallisto | bustools\n",
        "---\n",
        "Process the ClickTag reads with kITE (kallisto Indexing and Tag Extraction)\n",
        "\n",
        "We leveraged the lightweight and fast infrastructure of kallisto | bustools to quantify Feature Barcodes from single-cell RNA-seq data. Feature Barcodes are any type of single-cell biological information encoded with a barcoded DNA oligo, such as cellular proteins, sample identifiers, or antigen identities. Our kITE (kallisto Indexing and Tag Extraction) workflow is described in more detail in Gehring et. al 2019, but in brief kITE consists of generating a list of all Feature Barcode sequences as well as all of their Hamming distance = 1 mismatches. That list is used to generate a kallisto index with the k-mer length set to the length of the barcode. In this way, kallisto bus effectively searches the fastq files for the barcode sequences (and their single-base mismatches) and outputs a corresponding BUS file. Bustools can then be used to complete data processing. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lfI25BNhTU3m",
        "colab_type": "code",
        "outputId": "bb24059d-ddf6-41a4-c773-01d80ff393e8",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 867
        }
      },
      "source": [
        "\"\"\"\n",
        "The --header flag is used if the FeatureBarcodes.csv contains a header in the first row.\n",
        "\n",
        "The --t2g and --fa flags can be used to specify the name and location of the output files.\n",
        "The default filepaths are shown below as an example. \n",
        "\"\"\"\n",
        "\n",
        "!python3.6 ./featuremap.py ClickTagWhitelist.csv --header --t2g ./FeaturesMismatch.t2g --fa ./FeaturesMismatch.fa"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "CSV includes header row\n",
            "\n",
            "Feature Barcode Length: 11\n",
            "\n",
            "Read 20 Feature Barcodes:\n",
            "\n",
            "ClickTag1\n",
            "AGCAGTTACAG\n",
            "ClickTag2\n",
            "CTTGTACCCAG\n",
            "ClickTag3\n",
            "GAACCCGGCAG\n",
            "ClickTag4\n",
            "TCGTAGATCAG\n",
            "ClickTag5\n",
            "ACGCGGAACAG\n",
            "ClickTag6\n",
            "CGCTATCCCAG\n",
            "ClickTag7\n",
            "GTTGCATGCAG\n",
            "ClickTag8\n",
            "TAAATCGTCAG\n",
            "ClickTag9\n",
            "ATCGCCATCAG\n",
            "ClickTag10\n",
            "CATAAAGGCAG\n",
            "ClickTag11\n",
            "TCACGGTACAG\n",
            "ClickTag12\n",
            "CACTCAACCAG\n",
            "ClickTag13\n",
            "GCTGTGTACAG\n",
            "ClickTag14\n",
            "TTGCGTCGCAG\n",
            "ClickTag15\n",
            "ATATGAGACAG\n",
            "ClickTag16\n",
            "CACCTCAGCAG\n",
            "ClickTag17\n",
            "GCTACTTCCAG\n",
            "ClickTag18\n",
            "TGGGAGCTCAG\n",
            "ClickTag19\n",
            "ATCCGGCACAG\n",
            "ClickTag20\n",
            "CCGTTATGCAG\n",
            "\n",
            "The t2g and fasta files are now ready \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ymLInQBMT5_4",
        "colab_type": "code",
        "outputId": "d402d32b-2640-4639-bdf9-0004846b4903",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 153
        }
      },
      "source": [
        "\"\"\"\n",
        "Generate a kallisto index, setting the k-mer length k to the Feature Barcode length.\n",
        "\"\"\"\n",
        "\n",
        "!kallisto index -i ./FeaturesMismatch.idx -k 11 ./FeaturesMismatch.fa"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[build] loading fasta file ./FeaturesMismatch.fa\n",
            "[build] k-mer length: 11\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 680 contigs and contains 680 k-mers \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "OORjh2QfQD01",
        "colab_type": "code",
        "outputId": "32e59208-fbf8-4e55-fd61-b405b0c3621d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 289
        }
      },
      "source": [
        "\"\"\"\n",
        "Inspect the index\n",
        "\"\"\"\n",
        "\n",
        "!kallisto inspect ./FeaturesMismatch.idx"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[index] k-mer length: 11\n",
            "[index] number of targets: 680\n",
            "[index] number of k-mers: 680\n",
            "[index] number of equivalence classes: 680\n",
            "#[inspect] Index version number = 10\n",
            "#[inspect] k = 11\n",
            "#[inspect] number of targets = 680\n",
            "#[inspect] number of equivalence classes = 680\n",
            "#[inspect] number of contigs = 680\n",
            "#[inspect] Number of k-mers in index = 680\n",
            "#EC.size\tNum.targets\n",
            "1\t680\n",
            "\n",
            "\n",
            "#EC.size\tNum.kmers\n",
            "1\t680\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lr8EUbF4jya1",
        "colab_type": "text"
      },
      "source": [
        "`kallisto bus` is run on each set of fastq files. This function outputs a BUS file, a custom file type designed to store (pseudo)alignment information from single-cell RNA-seq experiments (Melsted et. al 2018). Later, the AnnData `concatenate` function will be employed to combine cells from the four pools generated in this experiment.\n",
        "https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btz279/5487510"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ienTKqNCc1XM",
        "colab_type": "code",
        "outputId": "3e720aa5-945c-42f6-8661-68f51c224c87",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 459
        }
      },
      "source": [
        "\"\"\"\n",
        "Call kallisto bus using the directory of the desired fastq files, the index to be used,\n",
        "sequencing technology used, number of threads, and output folder\n",
        "\"\"\"\n",
        "\n",
        "!time kallisto bus -i ./FeaturesMismatch.idx -o ./ClickTagsBig1 -x 10xv2 -t 2 \\\n",
        "./96S_TagsBig1_S1_L001_R1_001.fastq.gz \\\n",
        "./96S_TagsBig1_S1_L001_R2_001.fastq.gz \\\n",
        "\n",
        "!time kallisto bus -i ./FeaturesMismatch.idx -o ./ClickTagsSmall1 -x 10xv2 -t 2 \\\n",
        "./96S_tags_S1_L001_R1_001.fastq.gz \\\n",
        "./96S_tags_S1_L001_R2_001.fastq.gz \\\n"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[index] k-mer length: 11\n",
            "[index] number of targets: 680\n",
            "[index] number of k-mers: 680\n",
            "[index] number of equivalence classes: 680\n",
            "[quant] will process sample 1: ./96S_TagsBig1_S1_L001_R1_001.fastq.gz\n",
            "                               ./96S_TagsBig1_S1_L001_R2_001.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 10,591,381 reads, 10,311,181 reads pseudoaligned\n",
            "\n",
            "real\t1m11.530s\n",
            "user\t2m18.981s\n",
            "sys\t0m0.924s\n",
            "\n",
            "[index] k-mer length: 11\n",
            "[index] number of targets: 680\n",
            "[index] number of k-mers: 680\n",
            "[index] number of equivalence classes: 680\n",
            "[quant] will process sample 1: ./96S_tags_S1_L001_R1_001.fastq.gz\n",
            "                               ./96S_tags_S1_L001_R2_001.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 2,802,242 reads, 2,756,093 reads pseudoaligned\n",
            "\n",
            "real\t0m18.722s\n",
            "user\t0m36.123s\n",
            "sys\t0m0.293s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "R4KdQO1rJQHh",
        "colab_type": "code",
        "outputId": "6418222e-13e5-49ff-9f4d-a56f14a4a3de",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 459
        }
      },
      "source": [
        "!time kallisto bus -i ./FeaturesMismatch.idx -o ./ClickTagsBig2 -x 10xv2 -t 2 \\\n",
        "./96S_BIG2_tags_S1_L001_R1_001.fastq.gz \\\n",
        "./96S_BIG2_tags_S1_L001_R2_001.fastq.gz \\\n",
        "\n",
        "!time kallisto bus -i ./FeaturesMismatch.idx -o ./ClickTagsSmall2 -x 10xv2 -t 2 \\\n",
        "./96S_SMALL2_tags_S2_L001_R1_001.fastq.gz \\\n",
        "./96S_SMALL2_tags_S2_L001_R2_001.fastq.gz \\"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[index] k-mer length: 11\n",
            "[index] number of targets: 680\n",
            "[index] number of k-mers: 680\n",
            "[index] number of equivalence classes: 680\n",
            "[quant] will process sample 1: ./96S_BIG2_tags_S1_L001_R1_001.fastq.gz\n",
            "                               ./96S_BIG2_tags_S1_L001_R2_001.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 22,100,557 reads, 21,608,677 reads pseudoaligned\n",
            "\n",
            "real\t2m28.519s\n",
            "user\t4m49.458s\n",
            "sys\t0m1.932s\n",
            "\n",
            "[index] k-mer length: 11\n",
            "[index] number of targets: 680\n",
            "[index] number of k-mers: 680\n",
            "[index] number of equivalence classes: 680\n",
            "[quant] will process sample 1: ./96S_SMALL2_tags_S2_L001_R1_001.fastq.gz\n",
            "                               ./96S_SMALL2_tags_S2_L001_R2_001.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 2,022,440 reads, 1,972,226 reads pseudoaligned\n",
            "\n",
            "real\t0m13.665s\n",
            "user\t0m26.480s\n",
            "sys\t0m0.205s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ucpizrv1cWrw",
        "colab_type": "code",
        "outputId": "726dba3f-acab-4ba7-e350-6b9456dbc7d8",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!ls"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "10xv2_whitelist.txt\t\t\t ClickTagsBig2\n",
            "96S_BIG2_tags_S1_L001_R1_001.fastq.gz\t ClickTagsSmall1\n",
            "96S_BIG2_tags_S1_L001_R2_001.fastq.gz\t ClickTagsSmall2\n",
            "96S_SMALL2_tags_S2_L001_R1_001.fastq.gz  ClickTagWhitelist.csv\n",
            "96S_SMALL2_tags_S2_L001_R2_001.fastq.gz  featuremap.py\n",
            "96S_TagsBig1_S1_L001_R1_001.fastq.gz\t FeaturesMismatch.fa\n",
            "96S_TagsBig1_S1_L001_R2_001.fastq.gz\t FeaturesMismatch.idx\n",
            "96S_tags_S1_L001_R1_001.fastq.gz\t FeaturesMismatch.t2g\n",
            "96S_tags_S1_L001_R2_001.fastq.gz\t kallisto\n",
            "bustools\t\t\t\t kallisto_linux-v0.46.0.tar.gz\n",
            "bustools_linux-v0.39.3.tar.gz\t\t Mus_musculus.GRCm38.cdna.all.idx\n",
            "ClickTagsBig1\t\t\t\t transcripts_to_genes.txt\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_xc4ykPWnZa0",
        "colab_type": "text"
      },
      "source": [
        "After pseudoalignment with `kallisto bus`, bustools are used to count, sort, and error-correct the BUS file. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YuMRJzvhJh-r",
        "colab_type": "code",
        "outputId": "a7226814-4a3e-4157-8a88-5fd40ffdd18f",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!mkdir ClickTagsBig1/genecount/ ClickTagsBig1/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p ClickTagsBig1/output.bus | \\\n",
        "bustools sort -T ClickTagsBig1/tmp/ -t 2 -p - | \\\n",
        "bustools count -o ClickTagsBig1/genecount/genes -g FeaturesMismatch.t2g -e ClickTagsBig1/matrix.ec -t ClickTagsBig1/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x1a04000 @  0x7f7aee24a1e7 0x46c68d 0x46c709 0x45d143 0x43f308 0x7f7aed46ab97 0x443241\n",
            "Found 737280 barcodes in the whitelist\n",
            "Number of hamming dist 1 barcodes = 20550336\n",
            "Processed 10311181 bus records\n",
            "In whitelist = 10054978\n",
            "Corrected = 62662\n",
            "Uncorrected = 193541\n",
            "Read in 10117640 BUS records\n",
            "\n",
            "real\t0m31.162s\n",
            "user\t0m28.190s\n",
            "sys\t0m4.890s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "i2NWClORLeg_",
        "colab_type": "code",
        "outputId": "cf7771c9-d859-4449-83ee-b1c530d8eca9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!mkdir ClickTagsSmall1/genecount/ ClickTagsSmall1/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p ClickTagsSmall1/output.bus | \\\n",
        "bustools sort -T ClickTagsSmall1/tmp/ -t 2 -p - | \\\n",
        "bustools count -o ClickTagsSmall1/genecount/genes -g FeaturesMismatch.t2g -e ClickTagsSmall1/matrix.ec -t ClickTagsSmall1/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x1286000 @  0x7f6b8f3a11e7 0x46c68d 0x46c709 0x45d143 0x43f308 0x7f6b8e5c1b97 0x443241\n",
            "Found 737280 barcodes in the whitelist\n",
            "Number of hamming dist 1 barcodes = 20550336\n",
            "Processed 2756093 bus records\n",
            "In whitelist = 2691335\n",
            "Corrected = 18631\n",
            "Uncorrected = 46127\n",
            "Read in 2709966 BUS records\n",
            "\n",
            "real\t0m24.466s\n",
            "user\t0m23.115s\n",
            "sys\t0m3.753s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YwPFST7iLeey",
        "colab_type": "code",
        "outputId": "65519564-5167-4e02-cd7b-3cb23038c383",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!mkdir ClickTagsBig2/genecount/ ClickTagsBig2/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p ClickTagsBig2/output.bus | \\\n",
        "bustools sort -T ClickTagsBig2/tmp/ -t 2 -p - | \\\n",
        "bustools count -o ClickTagsBig2/genecount/genes -g FeaturesMismatch.t2g -e ClickTagsBig2/matrix.ec -t ClickTagsBig2/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x2eb8000 @  0x7f11520bc1e7 0x46c68d 0x46c709 0x45d143 0x43f308 0x7f11512dcb97 0x443241\n",
            "Found 737280 barcodes in the whitelist\n",
            "Number of hamming dist 1 barcodes = 20550336\n",
            "Processed 21608677 bus records\n",
            "In whitelist = 21060768\n",
            "Corrected = 156660\n",
            "Uncorrected = 391249\n",
            "Read in 21217428 BUS records\n",
            "\n",
            "real\t0m38.218s\n",
            "user\t0m34.116s\n",
            "sys\t0m5.850s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aWcVe4jvLeb-",
        "colab_type": "code",
        "outputId": "674db7b1-dbcb-4426-9f5e-3fa140d00b10",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!mkdir ClickTagsSmall2/genecount/ ClickTagsSmall2/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p ClickTagsSmall2/output.bus | \\\n",
        "bustools sort -T ClickTagsSmall2/tmp/ -t 2 -p - | \\\n",
        "bustools count -o ClickTagsSmall2/genecount/genes -g FeaturesMismatch.t2g -e ClickTagsSmall2/matrix.ec -t ClickTagsSmall2/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x1350000 @  0x7fbee6c8f1e7 0x46c68d 0x46c709 0x45d143 0x43f308 0x7fbee5eafb97 0x443241\n",
            "Found 737280 barcodes in the whitelist\n",
            "Number of hamming dist 1 barcodes = 20550336\n",
            "Processed 1972226 bus records\n",
            "In whitelist = 1925891\n",
            "Corrected = 13317\n",
            "Uncorrected = 33018\n",
            "Read in 1939208 BUS records\n",
            "\n",
            "real\t0m23.938s\n",
            "user\t0m22.720s\n",
            "sys\t0m3.687s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "O0uezY1_daZe",
        "colab_type": "code",
        "outputId": "47e3740e-8b0d-41c1-ea1e-1a79bc04f025",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 136
        }
      },
      "source": [
        "!ls -lh ClickTagsBig1/"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 315M\n",
            "drwxr-xr-x 3 root root 4.0K Oct 18 19:12 genecount\n",
            "-rw-r--r-- 1 root root 5.1K Oct 18 19:08 matrix.ec\n",
            "-rw-r--r-- 1 root root 315M Oct 18 19:08 output.bus\n",
            "-rw-r--r-- 1 root root  421 Oct 18 19:08 run_info.json\n",
            "drwxr-xr-x 2 root root 4.0K Oct 18 19:12 tmp\n",
            "-rw-r--r-- 1 root root 9.7K Oct 18 19:08 transcripts.txt\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MTKV6z1-WCFy",
        "colab_type": "code",
        "outputId": "b9bbdca3-b5ef-4962-a4d0-eb34a4ccdbed",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 102
        }
      },
      "source": [
        "!ls -lh ClickTagsBig1/genecount/"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 18M\n",
            "drwxr-xr-x 2 root root 4.0K Oct 18 19:11 genes\n",
            "-rw-r--r-- 1 root root 3.6M Oct 18 19:12 genes.barcodes.txt\n",
            "-rw-r--r-- 1 root root  211 Oct 18 19:12 genes.genes.txt\n",
            "-rw-r--r-- 1 root root  14M Oct 18 19:12 genes.mtx\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "herllm3J85lo",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Glance at the ClickTag data\n",
        "---\n",
        "We enjoy the ScanPy single-cell analysis package (Wolf, Angerer, and Theis; Genome Biology 2018)\n",
        "https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dyZJ07drdfUq",
        "colab_type": "code",
        "outputId": "abce4c49-4182-46e1-fc10-1427b22c2b5b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "!pip install leidenalg\n",
        "!pip install scanpy"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting leidenalg\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/b6/cc/d76baf78a3924ba6093a3ce8d14e2289f1d718bd3bcbb8252bb131d12daa/leidenalg-0.7.0.tar.gz (92kB)\n",
            "\u001b[K     |████████████████████████████████| 102kB 9.8MB/s \n",
            "\u001b[?25hCollecting python-igraph>=0.7.1.0 (from leidenalg)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/0f/a0/4e7134f803737aa6eebb4e5250565ace0e2599659e22be7f7eba520ff017/python-igraph-0.7.1.post6.tar.gz (377kB)\n",
            "\u001b[K     |████████████████████████████████| 378kB 22.9MB/s \n",
            "\u001b[?25hBuilding wheels for collected packages: leidenalg, python-igraph\n",
            "  Building wheel for leidenalg (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for leidenalg: filename=leidenalg-0.7.0-cp36-cp36m-linux_x86_64.whl size=1105355 sha256=daec44ff78243518d98d60a5018b220b562c9a339244e6614f4783bc228c160e\n",
            "  Stored in directory: /root/.cache/pip/wheels/29/55/48/5a04693a10f50297bcda23819ca23ab3470a61dd911851c8bd\n",
            "  Building wheel for python-igraph (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for python-igraph: filename=python_igraph-0.7.1.post6-cp36-cp36m-linux_x86_64.whl size=2218050 sha256=0c4f56fc498e023c9160ddf921b8347ba2a182063a8f819799fd8c0fc93db0e3\n",
            "  Stored in directory: /root/.cache/pip/wheels/41/d6/02/34eebae97e25f5b87d60f4c0687e00523e3f244fa41bc3f4a7\n",
            "Successfully built leidenalg python-igraph\n",
            "Installing collected packages: python-igraph, leidenalg\n",
            "Successfully installed leidenalg-0.7.0 python-igraph-0.7.1.post6\n",
            "Collecting scanpy\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/c9/df/e0957422a85e924501afbff2f9283e02d93cefcb04aa8319ade7af68d266/scanpy-1.4.4.post1-py3-none-any.whl (1.9MB)\n",
            "\u001b[K     |████████████████████████████████| 1.9MB 6.6MB/s \n",
            "\u001b[?25hRequirement already satisfied: seaborn in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.9.0)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.6/dist-packages (from scanpy) (2.3)\n",
            "Requirement already satisfied: statsmodels>=0.10.0rc2 in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.10.1)\n",
            "Requirement already satisfied: patsy in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.5.1)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.14.0)\n",
            "Requirement already satisfied: importlib-metadata>=0.7; python_version < \"3.8\" in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.23)\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.6/dist-packages (from scanpy) (5.5.0)\n",
            "Collecting numba>=0.41.0 (from scanpy)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/53/34/22b6c2ded558976b5367be01b851ae679a0d1ba994de002d54afe84187b5/numba-0.46.0-cp36-cp36m-manylinux1_x86_64.whl (3.6MB)\n",
            "\u001b[K     |████████████████████████████████| 3.6MB 45.5MB/s \n",
            "\u001b[?25hRequirement already satisfied: scipy>=1.3 in /usr/local/lib/python3.6/dist-packages (from scanpy) (1.3.1)\n",
            "Requirement already satisfied: scikit-learn!=0.21.0,!=0.21.1,>=0.19.1 in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.21.3)\n",
            "Collecting anndata>=0.6.22rc1 (from scanpy)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/2b/72/87196c15f68d9865c31a43a10cf7c50bcbcedd5607d09f9aada0b3963103/anndata-0.6.22.post1-py3-none-any.whl (47kB)\n",
            "\u001b[K     |████████████████████████████████| 51kB 15.7MB/s \n",
            "\u001b[?25hRequirement already satisfied: tables in /usr/local/lib/python3.6/dist-packages (from scanpy) (3.4.4)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.6/dist-packages (from scanpy) (4.28.1)\n",
            "Requirement already satisfied: umap-learn>=0.3.0 in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.3.10)\n",
            "Requirement already satisfied: matplotlib==3.0.* in /usr/local/lib/python3.6/dist-packages (from scanpy) (3.0.3)\n",
            "Requirement already satisfied: pandas>=0.21 in /usr/local/lib/python3.6/dist-packages (from scanpy) (0.24.2)\n",
            "Requirement already satisfied: h5py in /usr/local/lib/python3.6/dist-packages (from scanpy) (2.8.0)\n",
            "Requirement already satisfied: numpy>=1.9.3 in /usr/local/lib/python3.6/dist-packages (from seaborn->scanpy) (1.16.5)\n",
            "Requirement already satisfied: decorator>=4.3.0 in /usr/local/lib/python3.6/dist-packages (from networkx->scanpy) (4.4.0)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from patsy->scanpy) (1.12.0)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.6/dist-packages (from importlib-metadata>=0.7; python_version < \"3.8\"->scanpy) (0.6.0)\n",
            "Requirement already satisfied: llvmlite>=0.30.0dev0 in /usr/local/lib/python3.6/dist-packages (from numba>=0.41.0->scanpy) (0.30.0)\n",
            "Requirement already satisfied: numexpr>=2.5.2 in /usr/local/lib/python3.6/dist-packages (from tables->scanpy) (2.7.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.6/dist-packages (from matplotlib==3.0.*->scanpy) (0.10.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib==3.0.*->scanpy) (1.1.0)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib==3.0.*->scanpy) (2.4.2)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.6/dist-packages (from matplotlib==3.0.*->scanpy) (2.5.3)\n",
            "Requirement already satisfied: pytz>=2011k in /usr/local/lib/python3.6/dist-packages (from pandas>=0.21->scanpy) (2018.9)\n",
            "Requirement already satisfied: more-itertools in /usr/local/lib/python3.6/dist-packages (from zipp>=0.5->importlib-metadata>=0.7; python_version < \"3.8\"->scanpy) (7.2.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.6/dist-packages (from kiwisolver>=1.0.1->matplotlib==3.0.*->scanpy) (41.2.0)\n",
            "Installing collected packages: numba, anndata, scanpy\n",
            "  Found existing installation: numba 0.40.1\n",
            "    Uninstalling numba-0.40.1:\n",
            "      Successfully uninstalled numba-0.40.1\n",
            "Successfully installed anndata-0.6.22.post1 numba-0.46.0 scanpy-1.4.4.post1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uG_O_2Zadn9I",
        "colab_type": "code",
        "outputId": "9a7395e8-38ee-4b67-ec96-b9d113e18892",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 119
        }
      },
      "source": [
        "import scanpy.api as sc\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import pandas as pd"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.6/dist-packages/scanpy/api/__init__.py:6: FutureWarning: \n",
            "\n",
            "In a future version of Scanpy, `scanpy.api` will be removed.\n",
            "Simply use `import scanpy as sc` and `import scanpy.external as sce` instead.\n",
            "\n",
            "  FutureWarning\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LrEkYf3bUPKf",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "tags = sc.read('./ClickTagsBig1/genecount/genes.mtx')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "o61-eS26fvxO",
        "colab_type": "code",
        "outputId": "7dd80988-b87f-47f6-c4e5-a027b6d9f7a1",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "tags"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 219694 × 20 "
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 27
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Y8J42bsCqRn-",
        "colab_type": "code",
        "outputId": "56ee94d8-e051-44a6-e435-de1b5b446b24",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 187
        }
      },
      "source": [
        "!head ClickTagsBig1/genecount/genes.barcodes.txt"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "AAACCTGAGAAACGAG\n",
            "AAACCTGAGAAGAAGC\n",
            "AAACCTGAGAAGGACA\n",
            "AAACCTGAGAAGGCCT\n",
            "AAACCTGAGAAGGTTT\n",
            "AAACCTGAGAATCTCC\n",
            "AAACCTGAGAATGTGT\n",
            "AAACCTGAGAATGTTG\n",
            "AAACCTGAGACAGACC\n",
            "AAACCTGAGACAGAGA\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "G4bxhxbvqXau",
        "colab_type": "code",
        "outputId": "b4bc5c07-082c-4491-a7d4-a11182866dee",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 187
        }
      },
      "source": [
        "!head ClickTagsBig1/genecount/genes.genes.txt"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "ClickTag1\n",
            "ClickTag2\n",
            "ClickTag3\n",
            "ClickTag4\n",
            "ClickTag5\n",
            "ClickTag6\n",
            "ClickTag7\n",
            "ClickTag8\n",
            "ClickTag9\n",
            "ClickTag10\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-PDglMDx9DkQ",
        "colab_type": "text"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WFn5j1ctf2Ep",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "tags.obs_names=pd.read_csv('ClickTagsBig1/genecount/genes.barcodes.txt', header = None, index_col = 0, sep = '\\t').index"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-SFz5TybrwGO",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "tags.var_names=pd.read_csv('ClickTagsBig1/genecount/genes.genes.txt', header = None, index_col = 0, sep = '\\t').index"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YL1NAPhnrqAo",
        "colab_type": "code",
        "outputId": "0256910a-9118-411f-ecdf-95f2986998c4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 187
        }
      },
      "source": [
        "tags.obs_names"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['AAACCTGAGAAACGAG', 'AAACCTGAGAAGAAGC', 'AAACCTGAGAAGGACA',\n",
              "       'AAACCTGAGAAGGCCT', 'AAACCTGAGAAGGTTT', 'AAACCTGAGAATCTCC',\n",
              "       'AAACCTGAGAATGTGT', 'AAACCTGAGAATGTTG', 'AAACCTGAGACAGACC',\n",
              "       'AAACCTGAGACAGAGA',\n",
              "       ...\n",
              "       'TTTGTCATCTGTCCGT', 'TTTGTCATCTGTCTAT', 'TTTGTCATCTTCATGT',\n",
              "       'TTTGTCATCTTGACGA', 'TTTGTCATCTTGCAAG', 'TTTGTCATCTTGCCGT',\n",
              "       'TTTGTCATCTTGTACT', 'TTTGTCATCTTGTATC', 'TTTGTCATCTTTACAC',\n",
              "       'TTTGTCATCTTTCCTC'],\n",
              "      dtype='object', name=0, length=219694)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "u5j5PML1r0HB",
        "colab_type": "code",
        "outputId": "01d75b07-bc4b-4ddc-e606-97d0bfb2508d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 102
        }
      },
      "source": [
        "tags.var_names"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['ClickTag1', 'ClickTag2', 'ClickTag3', 'ClickTag4', 'ClickTag5',\n",
              "       'ClickTag6', 'ClickTag7', 'ClickTag8', 'ClickTag9', 'ClickTag10',\n",
              "       'ClickTag11', 'ClickTag12', 'ClickTag13', 'ClickTag14', 'ClickTag15',\n",
              "       'ClickTag16', 'ClickTag17', 'ClickTag18', 'ClickTag19', 'ClickTag20'],\n",
              "      dtype='object', name=0)"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 33
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3ch1a_GxpBej",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.filter_cells(tags, min_counts=0)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0728927GpFuW",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.filter_genes(tags, min_counts=0)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XPX3E2RuoyIt",
        "colab_type": "code",
        "outputId": "d1eb7d27-b033-4196-fc33-b3449563c08c",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 262
        }
      },
      "source": [
        "sc.pl.violin(tags, keys='n_counts')"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaYAAAD1CAYAAAAbHFHuAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzs3XdcV9X/wPHXEREHKoIoOFBzhoZi\nWqZfTBMn7lzlyi2OzJEjLXcTM03NEZW5NRfuvTUVFQcOxIkDGTJkr/fvDz5+fpCgaH6E6jwfj/uI\ne+65557Ph+rNuffc91EigqZpmqblFLmyuwOapmmalpYOTJqmaVqOogOTpmmalqPowKRpmqblKDow\naZqmaTmKDkyapmlajqIDk6Zpmpaj6MCkaZqm5Sg6MGmapmk5Su7s7oApFC1aVMqWLZvd3dA0TdPS\nOHXqVIiI2D6r3r8yMJUtWxZvb+/s7oamaZqWhlLqVlbq6Vt5mqZpWo6iA5OmaZqWo+jApGmapuUo\nOjBpmqZpOYoOTJqmaVqOogOTpmmalqOYPDAppcyUUmeUUpsN++WUUseVUv5KqVVKqTyGcgvDvr/h\neNk0bYwzlF9RSjU1dZ81TdO07PMqRkzDgEtp9r8BZopIBSAM6GMo7wOEGcpnGuqhlHIEugBVgWbA\nPKWU2Svot6Zlu7t377Jz587s7oamvVImDUxKqVKAG/CzYV8B7wF/GKosBtoafm5j2MdwvJGhfhtg\npYjEi8gNwB94y5T91rTM7N69m7i4uFd2valTp/Lll1++sutpWk5g6hHTD8BoIMWwbwOEi0iSYf8O\nUNLwc0kgAMBwPMJQ31iewTlGSqn+SilvpZR3cHDwy/4cmgaAn58fsbGxr+x6ly9ffmXX0rScwmSB\nSSnVEggSkVOmukZaIrJQRGqJSC1b22emYtK0FzJo0CCKFCmS3d3QtH81U+bKqwe0Vkq1APIChYBZ\ngJVSKrdhVFQKuGuofxcoDdxRSuUGCgOhacofS3uOpmma9i9jshGTiIwTkVIiUpbUyQt7RaQrsA/o\nYKjWE9ho+NnLsI/h+F4REUN5F8OsvXJAReCEqfqtaZqmZa/syC4+BliplJoGnAE8DeWewBKllD/w\nkNRghoj4KqVWAxeBJGCwiCS/+m5rmqZpr8IrCUwish/Yb/j5OhnMqhOROKBjJudPB6abroeapmla\nTqEzP2iapmk5ig5MmqZpWo6iA5OmaZqWo+jApGmapuUoOjBpmqZpOYoOTJqmaVqOogOTpmmalqPo\nwKRpmqblKDowaZqmaTmKDkyapmlajqIDk6Zpmpaj6MCkaZqm5Sg6MGmapmk5ig5MmqZpWo5iyqXV\n8yqlTiilziqlfJVSkw3lvymlbiilfAxbDUO5UkrNVkr5K6XOKaVqpmmrp1LqqmHrmdk1NU3TtH8+\nU67HFA+8JyJRSilz4LBSapvh2Kci8sdf6jcndXXaisDbwE/A20opa2AiUAsQ4JRSyktEwkzYd03T\nNC2bmHJpdRGRKMOuuWGTp5zSBvjdcN6fgJVSyh5oCuwSkYeGYLQLaGaqfmuapmnZy6TPmJRSZkop\nHyCI1OBy3HBouuF23UyllIWhrCQQkOb0O4ayzMo1TdO0fyGTBiYRSRaRGkAp4C2lVDVgHFAFqA1Y\nA2NexrWUUv2VUt5KKe/g4OCX0aSmaZqWDV7JrDwRCQf2Ac1E5L7hdl088CvwlqHaXaB0mtNKGcoy\nK//rNRaKSC0RqWVra2uKj6Fpmqa9AqaclWerlLIy/JwPaAxcNjw3QimlgLbABcMpXkAPw+y8OkCE\niNwHdgBNlFJFlFJFgCaGMk3TNO1fyJSz8uyBxUopM1ID4GoR2ayU2quUsgUU4AMMNNTfCrQA/IEY\noBeAiDxUSk0FThrqTRGRhybst6ZpmpaNTBaYROQc4JxB+XuZ1BdgcCbHfgF+eakd1DRN03IknflB\n0zRNy1F0YNI0TdNyFB2YNE3TtBxFByZN0zQtR9GBSdM0TctRdGDStL8pdUKppmkviw5MmvY3nD59\nmqFDh2Z3NzTtX8WUL9hq2r9ezZo1qVmz5rMrapqWZXrEpGmapuUoOjBp2jPEx8fj7u5u8uucPXuW\n+Ph4k19H03I6HZg07RksLCz4+OOPTX6dHTt2EBISYvLraFpOp58xaVoWvP766ya/xujRo01+DU37\nJ9AjJk17Rdq3b09KSorJ2k9KSsLV1dVk7Wvaq6IDk6a9IvPmzSNXLtP9J5c7d242b95ssvY17VXR\ngUnTXhE7OzuTXyNv3rwmv4ammZopV7DNq5Q6oZQ6q5TyVUpNNpSXU0odV0r5K6VWKaXyGMotDPv+\nhuNl07Q1zlB+RSnV1FR91rR/g+vXr/PHH39kdzc07YWZcsQUD7wnItWBGkAzw5Lp3wAzRaQCEAb0\nMdTvA4QZymca6qGUcgS6AFWBZsA8w6q4mqZlwNLSEnt7++zuhqa9MJMFJkkVZdg1N2wCvAc8/nNu\nMdDW8HMbwz6G442UUspQvlJE4kXkBqlLr79lqn5r2j9dsWLFqFevXnZ3Q9NemEmfMSmlzJRSPkAQ\nsAu4BoSLSJKhyh2gpOHnkkAAgOF4BGCTtjyDc9Jeq79Sylsp5R0cHGyKj6Np/yonT57k7t27T61z\n5MgRYmNjX1GPNC2VSQOTiCSLSA2gFKmjnComvNZCEaklIrVsbW1NdRlNy5KIiAiWLl2a6fHo6Gjj\nz0lJSZnWeywlJYWzZ8++lL49FhgYSERExFPrnDlzhkePHr3U62ras7ySWXkiEg7sA94BrJRSj1/s\nLQU8/pPtLlAawHC8MBCatjyDczQtRzIzM8PS0jLT4+PGjcPX15fk5GQaN278zPYiIyNZuXLlU+uM\nHj2a0NDQLPexVatWODo6PrXOkCFDKFasWJbb1LSXwZSz8myVUlaGn/MBjYFLpAaoDoZqPYGNhp+9\nDPsYju+V1IVuvIAuhll75YCKwAlT9VvTXgZLS0vatm2b6fHZs2dTtWpVzMzM2Ldv3zPbs7Ky4quv\nvnqi/PLly8ZlN3r16kWRIkVeuM8rV67Ex8cnw2N169Z94XY17XmZcsRkD+xTSp0DTgK7RGQzMAYY\noZTyJ/UZkqehvidgYygfAYwFEBFfYDVwEdgODBaRZBP2W9NytMTERLy9vQGoUqWKMWC9/vrrf+sF\nXk9PTwoVKpThsSNHjrxwu5r2vEyWK09EzgHOGZRfJ4NZdSISB3TMpK3pwPSX3UdNy4qZM2fSrVs3\ncsqzy/DwcHbt2kWtWrUAnnrL8Hls2LCBAgUKZHgsdYKspr0aOvODpj2Dv78/Bw4cyO5uGNna2jJu\n3DhWr15NYGDgS2s3s6Ckaa+azi6uac8wd+7c7O5ChqysrMiTJ0+W6vr6+lK0aFGKFy9u4l5p2t+n\nR0ya9g/VpEkTrK2ts1T30qVL3Lt3z8Q90rSXQwcmTfsP6NChA87OTzzyzdTMmTNJSEgwYY80LXM6\nMGnaXwQFBWV3F7JdhQoVMDPTKSm17KEDk6b9xcCBA0lO/m+/kdCqVSsdmLRsoyc/aNpfrFu3zuTX\nSElJITExEQsLiwyPHz58mJCQEJP3Q9NyIj1i0rSXLDo6mkaNGj21zs6dO5k1a1amxytUqICTk1OW\nr3nu3Dk9uUH719AjJk17yQoUKMCePXuA1EX7SpYs+cTIqFmzZjRr1izTNp53tdsbN26QkpJCiRIl\nnr/DGUhISCAmJgYrK6uX0p6mPQ89YtK0FxAeHg7A3bt3nzp7bcGCBRw+fPhvXSsuLo5vv/32qXXa\ntGlDjRo1/tZ10vrzzz/x9PR8dkVNMwEdmDQtE4sXLyYoKIiYmJgnjg0aNIiwsDCWLl3K5cuXM23j\nvffey/Al2EuXLpGYmJilfuTJkwc3N7cs93vDhg1s3749y/UzUr9+fUaOHPm32tC0F6UDk6ZlomTJ\nkjx8+JCBAwc+cWz58uUUKVKEMWPGPPVZUNOmTXFxcXmifN26ddy/fx9InQjxtBFRrly5qFq1apb7\n/fbbb1OzZs0s19e0nEYHJk3LhKurK+vXr+e3337D09PzpS7UN378eBwcHIDUwFO5cuVM60ZHR3Pz\n5s0st21vb6/XUNL+0XRg0rSncHJyIleuXLi4uFC2bNksndOnTx+uXr36XNdp06ZNpsfi4+MJDg5+\nrvY07Z/MlAsFllZK7VNKXVRK+SqlhhnKJyml7iqlfAxbizTnjFNK+SulriilmqYpb2Yo81dKjTVV\nnzXtrx4/26lUqRKFCxfO0jnz58+nYsWKL60P1tbW1K5d+4XP9/X15euvv870eHJyMvHx8S/cvqa9\nbKacLp4EjBSR00qpgsAppdQuw7GZIuKRtrJSyhHoAlQFSgC7lVKVDIfnkroC7h3gpFLKS0QumrDv\nmvbCzM3Ns7sL6VSsWJEPP/zwifLAwEAWL16Mo6Mj/v7+DB8+PBt6p2lPMuVCgfeB+4afHymlLgEl\nn3JKG2CliMQDNwwr2T5eUNDfsMAgSqmVhro6MGn/SAkJCVleruJ5RUZGPrEKbZ48eYzPs9Kytram\nSZMmz5XcVdNehVfyjEkpVZbU1WyPG4qGKKXOKaV+UUoVMZSVBALSnHbHUJZZuaa9EiEhIbi7u7+0\n9saPH8/Jkydf6Nzo6GgGDRqU4bGYmBh69Ohh3F+wYIFxwkZYWBhXrlwxHnN1dSVPnjw6KGk5kskD\nk1LKElgLfCIikcBPQHmgBqkjqhkv6Tr9lVLeSilv/aBYe1mOHz/O3LlzGT9+/FPrLVu2LMvvDk2e\nPJkLFy5w586dLPcjPj6e27dvU6BAAYYNG5Zhnfz587NhwwbjfqNGjXjttdcAuHnzJsePHzceS1tP\n03IakwYmpZQ5qUFpmYisAxCRByKSLCIpwCL+/3bdXaB0mtNLGcoyK09HRBaKSC0RqWVra/vyP4z2\nn1SrVi2GDx9OqVKlnlqvSZMmvPPOO1lqMyUlBT8/vyxlhIiNjSUoKAh/f39WrFgBYAw2z1KhQgUK\nFiwIgLOzc7rRlKWlZbq6SUlJHDt2LEvtapqpmXJWngI8gUsi8n2acvs01doBFww/ewFdlFIWSqly\nQEXgBHASqKiUKqeUykPqBAkvU/Vb09IyMzOjUKFCJCYmsnjx4kzr2draZmnW3urVq9mwYQNfffUV\nXbp0eWb93LlzY2FhQdWqVRkzZgxRUVFPnVqeVZcvX2b06NHG/aioKPbv3/+329W0l8GUI6Z6QHfg\nvb9MDf9WKXVeKXUOaAgMBxARX2A1qZMatgODDSOrJGAIsAO4BKw21NU0k5o3bx7fffedcV8pxcmT\nJ5k7d+4Lt9myZUvatWuXrmzbtm3cvn07w/rm5ubpAp6lpSWbN2/OND/f6NGjOXTo0DP7UalSpXSB\nycrKinHjxmXlI2iayZlyVt5hQGVwaOtTzpkOTM+gfOvTztM0U4iOjjaOTszNzenRowcxMTHY29s/\n48wnffHFF0yZMoX8+fM/cczc3Jxcuf7/b8Rr165Rrly5dGVpbdmyhatXrzJixIgnjj1ObbR161Yc\nHByoVq1ahm3kypWLokWLPvfn0LRXQS97oWmZ+PTTT58oy58/f4bB5WlE5KnPn1xdXdPtL1q0iI8/\n/jjTJSxatWr11OtdunQJb29v3njjjXTlJ06cwMnJibx582ax55qWPXRKIk1LY9KkSezcuTPDY/v2\n7ePatWvP3eamTZvw9c363eevv/6aEiVKPDUbw759+zIsv3nzJsWLF6d169aULl063bGjR48SFhaW\n5X78VWRk5Aufq2nPQwcmTUtj4sSJNGnS5InypKQk4uLiSJ3T86S+fftmmmi1devWjBo1isuXL/PL\nL79kuS/dunXL8FmSiLB79+4nyu/fv4+7uzsWFhbUqFEDNzc3IiIijMcHDBjw1NuQ8fHx+Pn5ZXjs\n1q1bjBo1Kst917S/QwcmTUsjs8Czf/9+Ll68mOlUbQ8PD4KCgp4ov3v3/99ssLe3f66cd2vWrMkw\nQ4RSiunTn3gUS0JCAoMHD6ZAgQJA6rtKjydOzJo1i48++ogzZ84wc+bMDEdjd+7cYf369Rn2pUyZ\nMixcuDDLfde0v0MHJk3LAldX16cunJeSkmJcTj2tqVOnEhISwpEjRzh79iwzZ85MF6yeV+/evQkN\nDc3wWJkyZWjZsqVxP23Ovpo1a7J48WLOnz+Pn58fIgLADz/8YJwmXr58ecaMGfPCfdO0l0U9/hf0\n36RWrVri7e2d3d3Q/mUiIiKynGF827ZtXL582ZgY1c/Pj5SUFOzt7SlcuDARERGkpKRQpEiRTNsI\nCgqievXqFC1aFB8fH8zMzAgMDMTOzu6FP8P48eMZPXq08XOEhYWRP39+LCwsMqzfqVMnFi5ciJWV\n1QtfU9MeU0qdEpFaz6qnR0yalgVhYWEMGDAASJ2uPW/evEzrXrp0CWtr63Qr31aqVAkbGxtiY2OB\n1FuDO3bseOLcESNGGN9psra2pmjRokRGRjJz5kyA5wpKcXFx/PHHH+nKpk+fni64njlzhps3b9K6\ndesM2/D09EQpleHy8ppmKnrEpGnPKT4+nsTERGNan5MnT/LNN9/QrFkz+vbti4+PD9HR0dSpU4db\nt24Zn0tt3bqVhIQE2rZtm2nb9+7dw97e3visq0GDBgBUrlyZBQsWPFc/Y2JiWLVqFYULF+add95J\nN/EhKSkJf39/6tWrx4kTJ7CzszM+m/qr3377DRsbGyIiIggMDNSTILQXltUREyLy1A0oTmpqoW2G\nfUegz7POy87tzTffFE17GXx9fSUoKOiZ9UJDQyUsLCxd2c2bN2XChAl/6/rvvvuuvPvuuxkei4+P\nl/Xr10tcXJxcv3490za2b9/+xGfo37+/+Pr6SkxMTLry7777TuLj4zNsJyUlRZo3b/58H0DT0gC8\nJQv/D8/KrbzfSE0H9PhtPz/gk+cOlZr2D3Tu3DnjZIUaNWrw4MGDDOuZmZlx5MgR4/66deuws7Nj\n6tSpQOofgP379/9bfYmLi6NTp064uroSFRXFhAkTOHfuHJcuXeKnn34iMTExw4zlTZs2xdbWluXL\nl/PgwQPi4+O5ceMGjo6O5MuXL13d119/ndy5M37v/tq1ay91+Q9Ny0xWAlNREVkNpABIau66ZJP2\nStNygHv37tGlSxdq1KgBQKFChQgJCUlX5/79+/Tr14/NmzenezH34cOHJCf//38mvr6+REdHP3cf\nQkJCuHXrFgB58+Zl7ty57N69m8DAQMLCwhg7dizm5ubY29tz7do1fv7550zbKlq0KBYWFlhYWLB5\n8+YM67i5uWWaCilfvnxYW1s/92fQtOf2rCEVsB+wAU4b9usAB7IyHMuuTd/K0/6upKQkadmypURG\nRhrLQkJCMqzbrFkzuX37trRr1+5vXTMgIEAePXokKSkpxrKn3cr7q6tXr0rPnj1f6NqPHj2SK1eu\nyIABAyQpKUnOnj37Qu1o2tPwEm/ljSB1mYnySqkjwO/AUFMESU3LKWJjY3n33Xc5ffq0sWzYsGGE\nhYU98XLq8uXLKV26NN9++y329vaZ3u5r1KgRkPoi7IMHD/D09Ex3/JNPPqFjx45s3LiR+Pj4dKMz\nycIkpQoVKvDbb78BULduXTp16vTUd6YOHjzIF198AcD169eZNm0agwYNIjw8nNWrV5OSkkL79u0z\nzWSuaabyzMAkIqeBd4G6wACgqoicM3XHNC07TZw4kUuXLvHuu+8ay6ZNm8ayZcvw9PRk9uzZxvKf\nf/6ZjRs3MnbsWMqUKcPUqVM5ePBgupdVk5OTqVKlCgDu7u4EBgZSvnx5ABITEwkNDUUpxaJFi2jb\nti1169alcePGxMfHIyLGRQIfS0lJAWDQoEEEBAQAsGTJEuLi4gA4fPgwK1eupGTJksZzGjZsSHh4\nuHHfxcWFCRMmAODk5ET9+vVxcnLCxsaGJk2aMH36dAYMGMDYsWP//heqac/hmdnFlVI9/lJUUymF\niPxuoj5pWrabMWPGE2Vly5ZlyJAhxMTEpBvN1KlTh7Nnz+Lq6kqTJk0oVqwYlpaW1K9f31gnKCjI\nmGHc09OTlJQULl68CMDp06fZu3cva9asAVInGVSsWJEOHTowZ84cRITQ0FCCg4NJSkrC3t6eRo0a\nsX37dqZMmWJcvsLMzMw4zTyj50Rubm7pUi4ppdKlPOrbty979uzh6tWr9O/fHwcHB5KTk2natOkL\nf4+a9iKyciuvdprNBZgEZPw2XhpKqdJKqX1KqYtKKV+l1DBDubVSapdS6qrhn0UM5UopNVsp5a+U\nOqeUqpmmrZ6G+leVUj1f4HNq2kuTP39+HBwcjPuRkZEkJSVRpkwZ1q1b98Sy5ZA6cSDte0QREREs\nXboUgNKlS9O5c2eOHj1KZGQkZcqU4ZNPPuHtt98GUgPI4MGD2bRpE59//jmxsbHs27cPCwsLY1Aa\nNmwYLi4uT2RwGDFiBEePHgVg9uzZmU5seOzxLcBcuXJx7949fHx8jMeio6Np1qzZ83xVmvZCsnIr\nb2iarR9QE3jyv7wnJQEjRcSR1AkTg5VSjsBYYI+IVAT2GPYBmpO6nHpFoD/wE6QGMmAi8DbwFjDx\ncTDTtJzAzc2NpKQk/Pz8jJkd5s+fz+zZs0lJSeHXX3/l66+/pkKFCgB8//33JCcn8/XXXwPg7++P\nr68vly5dIjIykty5c3Pz5k0OHDhgvIZSis6dO1OvXj2++eYbAMLDw41BJyUlJcP1o77//nvq1q0L\npGasKFiw4FM/y+OZd9u2bcPGxiZdRojevXv/rdV7NS2rnjvzg1LKHLggIpWf87yNwBzD1kBE7iul\n7IH9IlJZKbXA8PMKQ/0rQIPHm4gMMJSnq5cRnflBy279+vXDw8OD1atXc/HiRYYPH862bdtITEyk\nZs2aXL58GSsrK9q3b5/h+evXr6dGjRr06tULSF1/6fFtuEGDBtG/f3+SkpJwd3fn5MmTiAiNGjVi\n7969L9znq1evcuDAAfr27cuSJUs4c+YM5ubmeHt7s3379nRJYTXtRby0XHlKqU1KKS/Dthm4AmSc\nGz/zNsoCzsBxoLiI3DccCiQ1swRASSAgzWl3DGWZlf/1Gv2VUt5KKe/g4ODn6Z6mGQUFBfH555+z\nevXqdOW//vqrMav3xx9/zJkzZzI8/9NPPyUmJoZZs2YZUwE9fl5z8OBBmjZtytKlS2nUqBEjR44k\nKioq3fkHDhwgISEBGxubdCvlpv0DskWLFmzZsoVatWpx8uRJIHVElVFQejyi+qvdu3en+wz379+n\nePHixmU5unfvzjfffEPjxo3p168fly9fzvQ707SXLSvPmDyAGYbtK6C+iGR5mo5SyhJYC3wiIumW\nwDTMa38pyfpEZKGI1BKRWra2ti+jSe0/KCwsjKNHj1KxYsV05SVLljQ+v/n+++9xdnbm3r17fPfd\nd8Y6SUlJbNiwATMzM9q1a8eJEyf49ttvuX79Onv27OHOnTukpKTQv39/SpQoQdu2bTl27Fi665w+\nfZqYmBjq16+Pubk5ly5dIiUlJV1gatmyJT169GDZsmWZfo4NGzbg6enJ+vXrjTP40rKxsUmXzHX6\n9OlERUVRvXp1Y5m5uTmurq68/vrrREZGcuPGjSx+i5r292TlGdOBNNsREXky50kmDLf91gLLRGSd\nofiB4RYehn8+Xl3tLpB2LehShrLMyjXtpatcuTKbN28mPDycY8eO8dVXXwHQpEkT46SG3LlzIyLY\n2Njg6upqPNfMzAwPDw8sLCz45ptv+P3335kyZQpubm7Y2Njg5ORE5cqVqVGjBubm5sycOZPGjRsD\nqYEkNDSU4cOHG5eYsLa2pmLFipw/f9747tTjxQgjIyPTjaj++q5RdHQ0IsLQoUONEx5+/PFHY5By\ndnbGzs7OmEZJRAgLCzO2n5CQwOLFiwkICKB69erExsZy7969l/hNa1rmMg1MSqlHSqnIDLZHSqnI\nzM5Lc74iNfnrJRH5Ps0hL+DxzLqewMY05T0Ms/PqABGGW347gCZKqSKGSQ9NDGWaZhL58uXD0tKS\nW7duGV9QnTBhAvv27QNg79691KxZEw8PD5ydnblz5w6JiYkopWjTpg3x8fFMmTLFmIE8ISGB7du3\n8+OPPwKpS16kpKRw4MABPDw8gMxfoM2dOzfVq1cnJiaGixcvGic43LhxwxiYAgIC6N27t/Gc5ORk\n4uLiaNu2Ld9++62xvECBAowfP964f+XKFT7++GMgdRRYtmxZPvkkNQ1m2skXV69e5fjx49SrV+9v\nfrOalkVZSQ/xIhvwP1Jv050DfAxbC1LTG+0BrgK7AWtDfQXMBa4B54FaadrqDfgbtl7PurZOSaS9\nDHFxcfLgwQMREZk9e7YcPHhQtm/fLvPmzZOAgAA5d+6ciIhMmDBBzp8/bzwvPj5eli5datxPTk6W\njRs3yvDhwyUuLk6mTZsmMTExkpCQIOHh4emu+fDhQxERuXDhgkybNk3eeOMNeffdd8XT01M6d+4s\nO3fulO3btxuvHR4eLhs2bDBe99atWxISEiIzZ86U2NjYdG0nJyfL6dOnJSIiQtq2bZvl7yE+Pl6u\nXbuW5fqalhleYkoiAJRSxZRSDo+3LAS8wyKiRMRJRGoYtq0iEioijUSkooi4ishDQ30RkcEiUl5E\n3hAR7zRt/SIiFQzbr1nts6b9HV5eXly4cIFDhw4xdOhQXFxcCAgIoHPnzpQqVYo33niDgwcPopQi\nMjKSffv24eXlRZ48eahXrx5hYWFA6suuFy9exMvLizNnzlC/fn0OHjyIubn5Eyvi9uzZk6SkJB4+\nfEipUqWMo6LOnTvz7bffsnPnTpYvX26cIffw4UPWrFlDeHg4hw8fZvLkyaxbt46UlBQuXbpEWFgY\nQ4YMMfbD2dmZQoUK8euvv9KxY0fjdZcsWUJ8fDze3t7cv3+frl27smrVKgYPHkyePHnYuXMnHTt2\n1KMm7ZXIyqy81kqpq8AN4ABwE9hm4n5pWrbz8vLC0tKSLVu2cO7cOeLi4ujbt2+6DNu+vr4MGTKE\nunXrUr58ec6cOUNMTAxjx45Nlz3i0KFDbNy4kTVr1lCmTBlee+01IiOfvCPu5eVF7ty5CQ8P57XX\nXiNv3rwkJCRgYWGBg4MDY8bL3gWHAAAgAElEQVSMwc3NjSpVquDp6UnJkiUZMGAA69ev5/z580ye\nPJljx44xYsQIKlasyHfffceIESOM7ffu3ZubN2/yxRdfcPHiReOzq4SEBK5evUq/fv3Ily8fnp6e\nODs707RpU5KSksifPz8TJ07k4MGDJvzGNc3gWUMq4Cypt9/OGPYbAp5ZGY5l16Zv5Wkvw5o1ayQ0\nNFRERCZNmiR+fn4iIjJ8+HDjz48dPHhQPv/8c+nQoYMcOHAgw/bi4+Nl1apVEhkZKT/++KO0adPm\nmX149913pVixYnLr1i25ffu2hISESPPmzSUoKEgWLVpkrHf9+nWZMWOGJCcnS0REhIikLuzXunVr\n6dGjh7Ger6+vJCcnS+3atcXDw0OCg4ONxwYPHixDhgyRTZs2pevD+vXrxdHRUX788UcZOnToM/us\naZnhJd7KSxSRUCCXUiqXiOwDnr00rqb9w+XOnZvt27ezfPlyPDw8OHHiBABjxoyhdOnS6V6OXbJk\nCT179qRgwYKcO3eOlStX8ueffxpnwSUkJHDmzBmuXbuGubk5ZcuWZcOGDVnqh52dHRMnTmT37t3Y\n2NiwcOFCbG1tsbe358svv6RBgwZcunSJESNG4OLiQqFChTh+/DhXr16ladOmfPfdd8b3shwdHYmL\ni+PEiRM0b96cCxcuGK/z4Ycf0qhRI5YtW2YczQUGBvLw4UOioqJo3769XlZdezWeFblInaBgSWrG\nhhXALOBoVqJedm16xKT9HTExMRIVFSUiIuPHj39iBFGvXj05cuSIBAUFyeXLl0VEZM+ePfK///1P\ngoKCJDk5WVJSUuTzzz+XiIgI2bFjh/z0009Sp06ddO1ER0dLcHCwdOnSRX788cd0o5fHHq/HlHYi\nQ/369SUuLk6io6MlLCxM7t+/b1zW/dSpU+Lr6ytt2rSRlStXyubNmyUsLEx++eUXEUldU6pDhw4i\nIjJz5kz59NNPpWvXrnLw4EE5fPiwuLi4iIhIbGys+Pr6yoQJE6R+/fpy9OhRGT16tFy9evVlfMXa\nfxRZHDFlJTCNJzXTghmp07s/Bmyy0nh2bTowaX/HqlWrZPHixRIcHCyjR4/OsM7UqVMlMjJS6tev\nL9OmTZMOHTpIpUqVJC4uTpKSkmT16tUiIrJ582aZOnVqunMfLwQ4btw4ef/99+X69evGAPJXLi4u\nUqFCBVmxYoWEh4dLaGiotGjRwtjOtGnTjHXHjx8vc+bMER8fH2NZYGCg9O3bVyIjI6Vv377pFiFM\nSkqShIQECQsLky+++ELGjh0rhw8fFpHUW4Pfffedccafpr0MWQ1MWbmVlxvYSepKtgWBVZJ6a0/T\n/nUOHjxIUFAQPXr0wNramq5duxrf7Ulr3LhxFCxYkC1bthAXF8ebb77JlStXMDMzY9asWQQFBREX\nF4eHhwd2dnbpzm3WrBn37t3jyy+/5I8//qBcuXK4ublhZWXFb7/9Rnx8POvWrWPlypWICI8ePUJE\nSEpKIk+ePAwYMIBTp06hlKJ8+fLGtaHGjh2Lu7s7lSv/fxrL4sWLM336dAoWLMj48eMpX748169f\nZ+3atQQHB2Nubo6VlRXm5uZ89dVX/PHHH/z555/4+/uzd+9emjVr9vgPVAD27Nljom9e0/5fVjI/\nTBaRqsBgwB44oJTabfKeaVo2ePDggTGLQq5cuXBycqJHj/RLku3du9e4yJ+lpSW9evUyLgeRK1cu\nihYtipubG5999hlffvnlE8tgtGrVKsPcc48Dj1KK+/fvc/r0aU6dOoWVlRUdOnTg9OnTzJ49mzVr\n1hATEwNAu3btqFatGgChoaF07NiRYcOGARAVFcWuXbsoVqyYcRr6l19+yYULF1i2bBlDhgwhOTmZ\nU6dOkStXLpYtW0b79u2pU6cOTk5OjBw5kqlTp7J//362bt2Ku7s7a9euTReoNM0ksjKsMvyLaEfq\nkupHgHNZPS87Nn0rT3tRp0+fFmdnZ6levbokJibKqFGj5M6dO089p2/fvrJw4UKZM2eOnD59Wk6c\nOCGenp5y+PBh2bNnj7Rr106io6NFRCQ4OFgSEhLSnX/37l3Ztm2bWFpayh9//GEsnzVrlvEF27Tn\nfPXVV+Lh4SFDhw6VmJiYdG1FR0fLkCFDJDY2VoKDg43PlhITE0Uk9dbikSNH5NatW8ZzEhISJCgo\nSG7fvi0PHz6UkSNHir+/v7G/ERERMn/+fONn0LQXxcu6laeUGqSU2k9qtgYboJ+IOJkuVGpa9smV\nKxdz586lRIkSnDlzhsGDB2Nvb8/Vq1dJTEzM8JwaNWpw7949Ll68yNGjR6lduza9e/emYsWKnDp1\nij59+pA/f34CAgIYO3Ysbdu2TXd+SkoKERERtG3blgYNGhjLu3fvbsyb9+mnn9KvXz8Aunbtypo1\na2jQoAH58uVj/vz59OjRg6ioKPLnz8+AAQO4f/8+o0aNMi6bkTt36mLVbm5uAOmWhjc3N8fW1pbS\npUuTL18+PvroI1577TUg9fZgbGwsderUYeTIkS/hG9a0Z8vKM6bSpGYGryoik0Tkoqk7pWnZpXr1\n6kRHR1O7dm0WLVpE165defToEStXruTOndT8xY8ePWLq1KmcPXuW2NhYrKyscHFxwcHBwRhIIPU5\nVPPmzY2JUUuXLo2vry8NGzZMd80DBw5gYWHBkiVLsLGxMZYXKVLEmIC1UaNGXLhwge3bt1OgQAHa\ntGljnK7+9ttv4+Ligo+PD0FBQXz22WfY2NiwYMECIHVBwSNHjhgXMaxbt266rOiOjo6cP38egLNn\nz7J7926UUuzbt4+SJUtSqFAh4/fyuA1NM6WsPGMaJyI+z6qnaf8Gixcv5rvvvqNq1arUrVuXwYMH\nU7hwYdq3b8+6dakJ8vPnz0+RIkWYMWMGQ4cOpXLlylSrVg0HBwd2797NvXv3eP/99/H09KRatWrG\nUQtAly5dKFasGF5eXixfvhyA1157LdPRGKQuw960aVNWr16Nl5cX3t7eODs7M3z4cACmTZtGREQE\nx44d44cffqBUqVIEBARgYWGBn58fgwYNYuvWrTx69Ahvb28GDhyIiBAUFMTXX3/NsmXLqFSpEpAa\n5B5P9mjYsCHNmzendevWxMfH88Ybb5AvXz6TfO+allaWc+Vp2n9B1apVcXZ2ZtOmTXTu3JlWrVoB\n8M0331C9enVWr16NmZkZISEhDB06lIMHDzJ58mSuX7/OBx98wPfff09kZCQ//vgj3t7e2NjY8OjR\nI2P7w4YNo0ePHgQGBhpfYi1VqhQbN24kKSkJSE1z9HhRQhHh/v37iAilS5dmwIAB7NmzhxMnTnDg\nwAHOnDnD2rVrGTVqFGXLlsXJyYn4+HjmzJlDVFQUefLkYe/evSQlJbFt2zbMzMwoV64cu3fvxsrK\nCldXV5ydnY1rTf2Vj48P8+bN47333qN06dIZ1tG0ly13dndA03ISZ2dnypQpw5w5czhy5Ahffvkl\n5cqV4/fffwdSl4eA1GnliYmJ+Pn54efnZ1yJ9s6dO4waNYozZ84wcOBAOnfujJmZGZB6Sw3g1KlT\nhIWFMWDAAEaMGMH333/P0KFDjc+BfHx8CA4O5vr16zx69IiwsDD8/PxQSrFixQqioqKoX78+5cuX\nZ8CAAcaMFBMnTmTFihUEBQUZl7OwtLTkk08+oXDhwrzxxhuEh4czZswYtm7dysyZM3n//feZNWsW\nBQoUwMLCgsqVK/PWW28BcOHCBaKiovj1118ZOXIkrVu3fkW/Be2/TgcmTUvDzMwMW1tbJk+eDMD5\n8+fTZQAfMWIEIkJiYqJxtdc9e/ZQvnx5kpOTmTNnDtu3b2fGjBmUK1eOFi1a4OfnR/Xq1dm1axe5\ncuWibdu2WFtbM3LkSN577z1WrFjBBx98QFJSEr/99ht9+/bl119/xd3dnbNnz5I/f366detG7dq1\ncXFxITY2lkaNGpE3b14aNWpEnz59GD58OB9++CGlS5c2PpeKjY1lzZo1DBo0iMDAQHLnzs3Jkyc5\nevQon332GY0bNyYxMZHGjRtTuXJl7t27R79+/di+fTs1atSgSpUq2NjY0K5dO44fP07jxo0pWLDg\nq/+laP85JruVp5T6RSkVpJS6kKZsklLqrlLKx7C1SHNsnFLKXyl1RSnVNE15M0OZv1Iqy0u6a9rf\ntX//fqpXr0737t2B1PeMWrduTXJyMq+99hqdOnUiKSmJXLlyUbt2bczMzHB3d+f27dscPXoUNzc3\nVqxYwdq1a0lKSiIsLIz3338fMzMzXnvtNTp27EjDhg3x8Ul9hBsWFmacBPH+++/z5ZdfkpCQQL58\n+Zg8eTI///wzPXv2ZODAgYwbNw4fHx/s7OwoWrQot27donv37uTPn59cuXLRrl07Fi9ejLm5OYUK\nFaJSpUrGbOW1atVCKYW5uTkPHjxg0qRJDB8+nIIFCxpz4c2dO5dx48Yxe/Zsfv/9d8qUKYOFhQVT\npkzJtt+H9h+SlTnlL7IB9YGawIU0ZZOAURnUdSQ1i7kFUI7UxQLNDNs14DUgj6GO47Ourd9j0l7U\n9evXje8SnT9/Xpo3by4nTpyQhIQEcXNzk4oVK0pERIQsXbpUWrRoIdWrVzeem5iYKEOGDDHux8fH\ni7Ozs9y+fVuSkpJk+fLl4u3tbTweEhIi/fr1k9GjR0vDhg2lR48ecuvWLRk6dKiEh4fL3bt3pU6d\nOlK4cGH54osvpFu3bjJp0qQn+hwRESGzZs2SR48eyapVqyQ+Pl7c3d2lePHicu3aNfHy8pL79+/L\nkiVL5Keffnri/KSkJDl+/LgsWrRIdu3aJcOGDTNmJN+3b59MnDhR2rdvL6+//rr8/vvvL+271v57\neNkLBb5AwDsIPMxi9TbAShGJF5EbpK5U+5Zh8xeR6yKSAKw01NU0k7h+/bpx9l21atVYsGABK1as\nIDQ0lLNnz+Li4sKmTZto3bo1e/bswczMzJgJwcPDI90aTA8fPqRmzZqcPn0aMzMzPvjgA3755Rfj\n8a+++opmzZoxcuRIatWqxeLFi3FwcMDZ2Znk5GS6devGo0ePqFy5MmPGjGHcuHHUr1+fBQsWGDM/\nwP8vBX/p0iWSk5PJkycPw4cPp2/fvlhZWTF37lyuXLlCUlISAwcONJ7n7u7OxYsXMTMzw8PDgy1b\ntuDk5ERKSgolS5bkl19+4bPPPqNfv35MmDABGxsb47M2TTOl7JiVN0Qpdc5wq6+IoawkEJCmzh1D\nWWblmmYSj5d9ADh+/DilSpXi+++/x87ODm9vbx4+fMiJEyf46KOPiIuLo1evXqSkpJCYmMjdu3dZ\ntGgRt27dAiAgIIBt27bRqlUrjh07xsKFCxk4cCCenp7MmzeP+vXrs3r1aooVK8a3334LwLFjxyhR\nogTW1tb88MMPFClShJCQEEJDQ3F0dKRhw4YUL17cOFECUl+QrVatGhEREXzwwQcAFChQgAoVKmBt\nbc327duxtbWlXbt2+Pn58euvqYtAz5o1C0dHRw4dOkS7du1YtWoVxYoVY8qUKbz++uuYm5tTvHhx\nmjdvTmBgIBYWFuzcufNV/jq0/6hXHZh+AsoDNYD7wIynV886pVR/pZS3Uso7ODj4ZTWr/Ydt3LiR\n+Ph4wsLCEBGKFy9OixYt6NKlC2ZmZqSkpNCkSRPMzMzw9PRERLhx4wZTp05l9+7dODk5MWvWLOLi\n4hgwYADXr1/n008/pUWLFhw9epRz584Z3x96rESJEsbR0BdffMH58+dJTExkzJgxQOo7Ta6uruTJ\nkyfdeW+99RZnzpwxrlZ75MgR6tatC0BISAjr16/nzp07LFq0iMmTJyMizJ49m507d+Lo6Ei9evUQ\nEbZs2cL69esJDg5mx44dmJmZUbBgQcLCwmjWrJlxfSlNM6ms3O970Q0oS5pnTJkdA8YB49Ic2wG8\nY9h2pClPVy+zTT9j0l6W5ORkGTBggFy8eNFY9sYbb0jZsmXFzMxMbGxsJCUlRZKSkmTs2LEyaNAg\nKVq0qLi5ucn8+fOlQoUK0rlzZ+nTp49888030rt3bzl27Jj4+/tLTEyMdOrUSUTEuBZTt27dxNbW\n1rhC7ttvvy2VKlUyrqS7dOlSWb9+/RP9DAsLk5CQEFm8eLF07dpVmjRpIocOHRIREW9vb/H09BQR\nkWHDhhlz//Xr10927Ngh77zzjjx48EA6deok5cqVk9jYWNm6dau0a9dORESioqKMa0w1aNDAFF+z\n9h9BFp8xvdLp4kopexG5b9htBzyesecFLFdKfQ+UACoCJwAFVFRKlQPuAl2AD19ln7X/nvDwcKKj\no7Gzs6NHjx7kz5+fGTNmULp0aSZOnMjMmTMJCAhg06ZNuLm5oZQiISGBwMBASpQogY+PD0ePHiUw\nMJAPPviAsmXLUq9ePWrWrMmQIUPw9fUlNjaWIUOGsGrVKkaMGMHvv//OO++8Q/ny5fniiy9wcHAA\nUnP3JSUlGV+A7dq1K1OmTKFixYpUrVqVpk2bsm3bNubOnYurqytvvvkmCxcuJCYmhrfffptOnTpx\n9+5dChYsSEhICBMnTjQ+E2vWrBmjRo3Czc2NgIAA4/T3vHnz0rx5cxwcHAgODmbKlCm4urry9ttv\ns23btuz5pWj/KSYLTEqpFUADoKhS6g4wEWiglKoBCHATGAAgIr5KqdXARSAJGCwiyYZ2hpA6gjID\nfhERX1P1WdP8/f2ZMmUKJUqUwMnJibx58zJs2DAaNmzIkSNH+OCDD5g5c6Yxeeq1a9d499132bVr\nl/HZDUDHjh3x8vIiJSWF3r174+XlZXyudPv2bY4cOcKkSZPo0aMHjx494rvvvqNevXrGRKqP5cmT\nh9KlSxMeHs6WLVuIiYnh999/p2fPngD88ccf5MqVi/HjxzN16lTat2/PW2+9xbhx4zA3NyciIoLT\np08TFBTE9u3bOXLkCBEREXTt2hVbW1vGjh1Ly5Yt6dSpEx07dmTOnDkcOHCAd955h1y5cjF8+HCS\nk5MJCgrC19eXRo0avfLfifbfox7/9fRvUqtWLfH29s7ubmj/QPv372fZsmUMHz6cy5cvc+XKFXbs\n2MGmTZsICAhg/fr19O7dm169ehEXF8e+ffv48ssvcXJyonnz5syYMYPatWvz3nvvkZycTFJSEubm\n5ty7d49PPvmEGzducO3aNe7fv0/37t25dOkS8fHxjBo1yrgo4ahRo3j99dextrbGzs4OGxsb5s2b\nh5eXF4cOHaJ+/fqsWLECLy8vzp8/z6effsrJkycJDQ2lVKlSdO3ald69exMQEGB8Z8rKyoojR45Q\ns2ZNKlWqhI2NDQ8fPiQqKorTp09TqVIlHB0dGTRoEMeOHWPfvn1YWVkRFxfHhAkTsLCwoGPHjtSo\nUSO7f0XaP5hS6pSI1HpWPZ35QdPScHZ2plixYlSoUAFHR0fGjBlDrVq1+OCDD2jbti3jx4831lNK\nsWrVKsaNG2ecqXfz5k2GDh0KpGaR+Prrr4mMjMTPz4/169cbg0C+fPkYMWIEJUuW5NSpUxw4cIBN\nmzbx4MEDY0LXXbt20b9/f27cuMGAAQM4cOAAb731Fnnz5qVMmTLY2tpy9OhRQkJCKFSoEIUKFSIx\nMZFChQpRt25dUlJS2L17N2fOnCFv3rzs3buXgIAA9u/fz7p16xg6dChr165lyZIllCxZkh49eqCU\nYsmSJYwZM4a+ffty9+5d1q5di7e3N71792bNmjVPTLzQtJdNJ3HVtDQKFy6MlZUVPXv25OjRo7Rs\n2ZIiRYqwcOFCVq9eTdu2bXFycuKPP/5gwoQJbN++nbNnzzJjxgwcHR2JiYnh+vXrxvY6derEpEmT\naNy4MQA1a9bE0tISf39/Ro8ezb179/D29qZDhw506NCB6dOnM27cOADefPNNChYsSHR0NOPGjSM+\nPp4LFy7g4uJC165diYqK4qOPPmLv3r0AxMfHY25uzs8//8yUKVO4evUqBQoUYNGiRbi7u9OkSRNq\n1arFsWPHmDVrFjdv3mTlypUUKFCApUuXUqVKFR49esS8efMoWrQo69evx8vLi1GjRlGvXj2srKz0\nshfaK6Fv5WlaGt7e3lhaWlK2bFk+/vhj8uXLh5+fH6tXr8bHxwdnZ2fjdO169epx6NAhChUqxOuv\nv85HH33E/fv3qVixIr169WLQoEFYWlpSvXp1GjZsyJ9//om1tTXJyckcPXqU7t2787///Q9zc3Mu\nX75MqVKlqFatGsuXL2fPnj2UKlWKWrVqERoayrZt26hSpQqjR4+mdu3aVK1aFVdXV+zs7ChVqhRt\n27bFw8ODlStXcvjwYQYNGgRAt27d+OKLL1i+fDnh4eGMGjWKAwcOcOfOHcaMGcPVq1extLRkz549\nFChQABcXF4oWLcrWrVu5ffs2hQsXplWrVtSpU4cqVapQp04dY9oiTXte+laepr2AuLg4Fi9eTPv2\n7bly5QrW1tbGmWjh4eF4eXlRr149Ll++zLVr11i8eDEPHz7k3r17WFtbU6hQIdauXcvQoUMpXbo0\ns2bNYsuWLTg4OLB37168vb1p3Lgx48eP5+bNm6SkpBAfH4+HhwcAR48eZePGjfzwww9MnTqVwMBA\n4+KDffv2pXv37iQkJHDv3j1sbGzYs2cPefLkYd++fdja2pI3b950y1NYWlqSmJjI2bNn+eyzzyhZ\nsiT29vZEREQAULFiRQAaN25M8+bNqV+/Pps2bWL+/PkMHDiQpUuXkpyczIULF4iNjTUux6FppqRH\nTJr2F0OGDMHV1ZXLly9TtmxZXFxcWLBgATdu3KB27dq0bt2aiRMnUqJECa5du8a5c+eIjo7m/fff\np0iRImzdupXjx49z7949WrZsSceOHWnfvj22tra4u7uTmJjIoEGDiI2NpVWrVnTu3Jlz584RFRVF\nmTJl+OSTTyhfvjzOzs7UrVuXy5cv4+/vT6FChYwZHw4fPkznzp05cuQIZcuWBVKXy/Dz86NTp07G\nz+Lv74+lpSVbtmwhISGBGjVqUKBAAZycnIx1XFxc2Lt3Lx06dODWrVs8ePCA2bNnkzdvXm7dukXt\n2rU5duwYn3zyCW5ubmzZsuWV/j60fw89YtK0FzRnzhwAOnfuzN27d3F3d6dZs2ZMmTIFDw8PTpw4\nwZAhQ3B0dOTrr78mJSWFN998k3HjxjFjxgycnZ25efOmcWLDjh07SElJYfPmzXTv3p3k5GRj2qKF\nCxeSO3duIiIicHBwoF+/fqxZs4Zu3bqRL18+zp49S4ECBZgxYwbvvfcekZGRXL16lV69erFy5Uqu\nXLmCj48Pe/fu5dGjRzg5OfHWW29hZ2eHl5cXFSpUYMOGDezYsYM7d+4wd+5cY8qltm3bsm7dOg4d\nOkRgYCB16tShUqVKfPjhh6xevZqwsDBmzpzJypUrSU5O5uzZsxQrVizbfi/af4ee/KBpaYgIf/75\nJ8nJyVy+fJk333yTPXv28N577zF79mwsLS3p2LEjDx8+5KOPPqJMmTKULVsWT09PYmNjSUpKIjg4\nmHfeeQd3d3diYmLw8PCgWLFitGrVigYNGtCqVSuWL19OUFAQ/fv3Z9myZaxdu5amTZvy66+/Ehoa\nys2bN/Hw8KBChQpUqlSJMWPG0KhRI9q1a8fAgQOpVq0as2fPpmvXrsZURz4+PiQnJ1OuXDkmTpyI\nvb09rVq1ok2bNsZbdhcuXMDd3Z2aNWuyaNEi49pNFhYWVK1alfPnz+Pl5YWzszPnzp0jNjaW0qVL\nM2/ePLp168aDBw+y89ej/UfoEZOmpbFgwQK2bt3K/PnzsbS0xMHBgQULFhAVFUXVqlVZtGgRb775\nJjVq1KBcuXJMnjyZpk2bYmdnR8OGDalfvz5dunShSJEi9OnTh8WLF2NpaUlAQADXr19nyZIluLu7\nU79+fZYuXWqcgVetWjUCAgJITk5m2rRpxv40aNAAIF2OukKFCtGjRw/s7Oxo06YNd+/eJSIigk6d\nOtGrVy+uXr1Kp06dmD59Og8ePKBPnz6EhIRQpkwZ9u7dy8SJE7l27Rru7u78+eefVKlShTlz5lC1\nalVCQkLInz8/rq6u7N69m/bt22Npacm6desoXLiwcSq7ppmSHjFpWhqenp5s3bqVdevWUaFCBaZN\nm8apU6cYMGAAbdu2NSZHnTt3LsOGDeP27dscOnQIpRTff/89jRs3pmPHjvz8888UL16c2NhY3njj\nDWrUqIGjoyMODg7Exsby+eefY29vz44dOwgNDeXAgQMUKlQIc3Nz5s+fD6SO3oKCgoiJiaFFixb8\n73//49SpU+zduxc7Ozusra05ePAgx44dY8OGDdy7d4/9+/fj4OCApaUlvXv3pmXLljx69Ij333+f\nGzdu4OHhweeff07hwoX55Zdf6N+/P7t27aJPnz5MnTqVqlWrYmZmxqlTpzh58iS2trb07NmT7du3\nU6dOHUqW1Mn9NdPTkx807S/Cw8NZtWoVQUFBzJ8/n2bNmhEfH0+JEiVwd3dny5YtiAhDhw7F3d2d\n3LlzY2dnR7Vq1YiOjiY+Pp5evXrxzTffoJSiZMmSeHl5sWDBAvLly8eMGTMYOXIknTt3Zu7cuXz+\n+ecEBQUxadIk5s6di6+vLzdv3kQpRf78+cmTJw/BwcHY29uTkJDAjBkzcHBwoGjRouzYsYO8efPy\n+++/079/f/7880/j+1dhYWFMnjyZwMBAypUrx/+1d9/xUVV548c/Z2YymZn0CiQBEkIngNKrIOpD\nE0SBRVAEG64/y8LaVlxFcV1A2X1EsVAeVJplRWkakaaI9BZ6SKgJCZn0kEyfOb8/krlOEFdXwcBy\n3q/XvObOnduSS+6Xc873nFNRUUFwcDAdO3YkKiqKJ554Ar1ez5AhQwDo1asXX331FadOneKhhx4i\nNTWVc+fO4Xa7SU9PJy0tjc2bN1O/fv06vkPK1eqXJj+oEpOiXGDbtm2UlJTwxz/+keXLl5OSksKu\nXbsYNGgQw4cPp2vXrgO29y0AACAASURBVAwbNox33nmHrl27kpWVRc+ePenfvz8tWrSga9euSClp\n1KgRRqOR3r17M3HiRAYMGMCDDz7Ixo0befXVV7nrrru45557aNiwIQUFBbz00kuUl5dTUFDApk2b\nyMnJoaysDCEEixcvpkWLFtx4443MmDGDffv2ER4eTosWLUhLS2PkyJFERUVRUFDA66+/zvbt2+nc\nuTNms5mJEyfi8XgICwvjuuuuY86cOSxevJh+/frRqlUrbZqNoUOHsmjRIpYsWUJOTg4hISGkpaVh\ntVp57bXX6NGjB7Nmzarju6NcC/QvvvhiXV/DJTd37twXJ0yYUNeXoVylUlJS+Prrr3n44YdJT0/n\n8ccfp127dsTExGA2m/F6vTz33HMcPHiQWbNm8eSTT5KYmEheXh5BQUHodDpsNhtjxowhNTWVr776\nihtuuIFhw4bRq1cvSktLadKkCWVlZXi9Xm688UaOHTtGeXk506dPJzs7m/3795Odnc2BAwc4deoU\nNpuNWbNmkZWVhd1ux+FwsGPHDvbt20dCQgIpKSn07NmTNm3acOedd2Kz2ejWrRvx8fG0adOGOXPm\n0K9fPywWCwaDgRUrVnD8+HGWLVvG2bNncblcpKenM2rUKJ555hmSk5N58MEHyc/PJz4+nuTkZJKS\nknjyySfR6/V1fYuUq9RLL72U/+KLL879ue1U8oOiBPj6668pLS3FarXSs2dPvv32W+666y70ej1P\nPfUUZrMZs9nM9ddfz+DBg8nIyOCpp55iyJAh7N69G5PJhM1mo2PHjjRu3JguXbpgs9nYuXMn9957\nLwA2m42RI0diNBpxOBwsWrSIe++9lx49ehAdHc2NN97IqlWrKCwsJCYmhpCQEBo3bszSpUt5/PHH\nmTJlCo888gijRo1i0KBBrF27ln79+pGXl0dGRgZPPPEE5eXlrFixgk8//ZTmzZuzcOFCIiIiWLp0\nKd999x2FhYV069aNzp074/V6eeCBB0hMTGTfvn1akHv++edp06YNx44d44EHHiAzM7PWzLmKcrmo\nqjxFCdCpUyf69OmDx+Ph/Pnz5Ofns2PHDj7++GM+/vhj/vd//5c2bdoQEhLC6dOnufnmmzl9+jTf\nf/893377LZGRkWzYsIGzZ88CUFpaSlRUFFlZWbz//vuMGDGCr776ihMnTmAwGAgNDaW0tBSv10vb\ntm1p2bIllZWVZGRksHv3bqSUuFwuUlNTycjIYNWqVWRlZdGkSRN27dpFv379cLlcxMTEEBwcTHBw\nMAaDgYMHDxIXF4cQgo4dOzJhwgT69evHpEmTKC4uZvz48URGRrJt2zZ69OjBoUOHyMvL47PPPuPw\n4cMUFxczefJkDhw4wIABA+jTpw+ZmZkIIer4DinXAhWYFCVAdHQ0lZWVZGZmEhYWxtSpUzl79iyD\nBg3CbrdTWVnJHXfcwX333ceAAQPo3r07H3zwAcuWLePzzz+nZcuWtG7dmnXr1jF79my2bNmiBRa3\n202bNm0YOnQoLVu2JDMzk5kzZ9K4cWNCQ0M5fvw4//d//0d6ejrJyclERUWxb98+cnNz6dOnDw0b\nNmTp0qWMHj2ayspKBg0aRGxsLPv27SMvLw+n08mhQ4cwm8307duXiIgIGjZsSM+ePVm/fj3PP/88\nAH/729/Iz8/n+PHjhIaGAtWluP79+2M2m2ncuDEFBQUcO3aM66+/nszMTCZMmKBNUuj1euvyFinX\ngMsWmIQQC4QQViHEwYB10UKItUKIrJr3qJr1QgjxhhAiWwixXwjRIWCfcTXbZwkhxl2u61UUgHPn\nztG6dWvatWvHkCFDSE9Pp0ePHqSmppKTk6P15xk1ahTNmzenefPm3HDDDSQkJDBu3DhatmzJo48+\nitVqZejQoTgcDr766ivmz5/Pxx9/zOLFiwkODubQoUMMGDBAG9euVatWVFRUEBkZyfTp01m5ciUm\nkwkhBNHR0Qgh2LlzJyUlJfz5z3/G4XCwcuVKCgsLcTqdlJSUaOPthYeH06BBAzweDx06dGDTpk3c\nfffdJCQkkJGRwenTp+nXrx/z58/HarXy0UcfMWrUKDIzM2ndujUul4vc3Fzuu+8+nn32WVJSUti9\neze7du2iW7duqo1JuewuZ4npfWDABev+AqyXUjYD1td8BhhI9XTqzYAJwDtQHcionvm2K9AFmOIP\nZopyOezcuZPExERyc3M5evQob775Jl999RUvv/wyTZo0YdWqVbz99tts3LhRm47822+/pby8nKlT\np+Lz+XjhhRfYvXs3M2bMYNiwYWzfvp39+/fTtGlTmjRpQnp6OhaLBZ1OxxtvvMFDDz3E0KFDee65\n55g5cyZjx45l9OjRNGrUiJ49e1JcXIzNZuPLL7/k1ltvZfz48Zw5c4ZTp06RlJTEbbfdxtixY9mw\nYQPHjh1jwIAB3HLLLRw5coTKykq2b9/O0KFD2bt3r9YW5u8A3KRJEw4dOoTX66Vhw4YIIUhISGDN\nmjVkZ2eTmprKm2++SVFREYsXL9bG5VOUy+myBSYp5Sag5ILVtwEf1Cx/AAwLWL9QVtsGRAohGgD9\ngbVSyhIpZSmwlh8HO0W5ZMrLy0lMTKSqqopbbrmFSZMm8eyzz/L9998zcuRICgoKWLFiBeHh4Ywd\nO5Y2bdoAMH78ePbu3cuECROYOnUqAO+++y5paWmUl5fz4osvkpeXR+/evXn11Vc5ePAg7777LkeP\nHqVbt24YDAZmzJjBokWLmDFjBnq9nvr162O329HpdBiNRsrKyqisrKRevXqcOnWKqqoqGjduTN++\nfZk4cSJ5eXk4HA7tZxk9ejSvv/46MTExpKWl8fDDD9OvXz8aN25MTk4OH374IUuXLmXOnDlkZ2ez\natUqLBYLJSUl7N69m5MnT+JyuVi0aBFLly5lyJAh/P3vf6+T+6JcW37vNqZ6Usr8muVzQL2a5UQg\nJ2C73Jp1P7X+R4QQE4QQu4QQuwoLCy/tVSvXjLvvvptZs2aRnJzMm2++SVRUFBaLhejoaAoKCtiw\nYQNz5swhNjYWi8XCtGnT2LlzJ506dcLhcLBixQr69OnD0KFDyczMpF+/fvTv35+RI0dqAaZdu3ZY\nrVZWrVpFZmYmgwcP5o477mDs2LH84Q9/oF27dpSVlbFmzRry8vJo1KgRZ8+exWQyMWDAAKKioliz\nZg2zZs2ic+fOXHfddXz22Wc0btyYbt26cd9993HixAlWrVpFz549OX36NDqdjkOHDnH+/HkqKiqY\nPXs22dnZOJ1OOnfujE6nIyEhgU6dOnHnnXdy9913s3//fpYvX8706dNZvXo1xcXFZGdn1/UtUq4B\ndZb8IKuHnLhkw05IKedKKTtJKTvFxcVdqsMq15jKyko+/PBDioqKqKioQKfTkZ2dzfLly0lNTeWF\nF16gsrKSsLAwLBaLVi03adIkmjZtSlVVFf/617+466672LlzJ//61784cuQIY8aMYf/+/fz973/n\no48+onXr1hw7doxZs2Zx5swZcnNzSUpKokOHDkgpycnJISgoCK/XS25uLkuWLCE1NZXExETeffdd\nJk+ezGeffYbVatVGmzCZTJSWlrJu3TruvPNORo0axVNPPYXNZiMlJYXS0lImT55Mp06d2LZtG82a\nNaNt27aMHz+eXr16MX78eBYtWsRLL73EhAkTGDBgAGvXruWdd95hxowZrFu3jg8++ODnf4mK8hv9\n3oGpoKaKjpp3a836s0DDgO2Satb91HpFuSw++ugjKisrefzxx0lJSSEkJASv10tycjIzZ86ke/fu\nxMTE0LFjR8aMGcOQIUO4++67iY+Pp0mTJnToUJ23k5GRwaBBg9i2bRuDBg2iefPm3HPPPaSlpfHE\nE08wYsQIDh06RIcOHejYsSNt27Zlz5492tTrt956K5GRkQghSE1NJTMzk5KSEgYMGMBrr73G3Llz\nadKkCefPn+evf/0rY8aM4ezZsyxYsICxY8eyYMECXnvtNUwmE7GxsVRVVTF69Gjt2p9++mlmz57N\nZ599xiOPPMI999xDdHQ0J0+epFu3brz99tucPHmSefPm0bJlSw4cOMAnn3yi9cVSlMvp9+4ttxIY\nB0yveV8RsP5RIcRHVCc6lEsp84UQa4C/ByQ8/A/w7O98zco1ZNSoUXzxxRcEBwfTsmVLhg4dyrx5\n89i9ezcnTpwgLy9PG6anpKSE2NhY9Ho97du3Z+DAgfTq1YvOnTtTXFyM2WxmxIgRpKWlUb9+fVwu\nFwDjxo0jKCiI9PR09u3bxzvvvMPatWtZtmwZq1atwuPxcPLkSdq3b8/27duJi4vj5ptvZty4cfTt\n25fvvvuO2NhYnnnmGSZOnMjo0aP5xz/+oQ1P9NZbb1G/fn2effZZXnvtNW3Q1ieeeILY2FjOnDnD\nxo0bOXPmDFJKZs6cqU3F3rhxY7788ksaNWrE3/72N9q2bcvcuXP5/vvvOX78OG+88UZd3h7lGnE5\n08U/BLYCLYQQuUKI+6kOSLcIIbKAm2s+A3wJnACygXnA/wOQUpYALwM7a15Ta9YpymVRXFzM1q1b\nWbRoEcOHD+eZZ56hsrKSxMREevfuTWpqKlarlbZt2/L4449jtVo5d+4cn376KV9++SXTp0+nrKyM\nl156iZYtW9KwYUP+9a9/Ua9ePZ577jmioqIICwtj/PjxhIaG8sYbbyClZMaMGXz11VckJiaSmJhI\nQUEBa9asoUuXLvh8PgYPHkxCQgJLlixh69atLFiwgEWLFlFcXMwXX3yB1Wpl165dxMXF8cADDzB3\n7lzeeustysrKuOWWW6hfvz5r167lD3/4A926dWPo0KHa8EbLly9n4cKF6PV6LXhu2rSJhQsXsnLl\nSpo1a8bzzz+vjUauKJebGl1cUQI4nU7mzp3Le++9R3BwMLfddhsvvPACTz/9NJGRkSxevBiDwcD5\n8+ex2WzcfvvtfPDBBzzwwAMcOXKEkpISbDYbOTk51KtXT6sKtNls9O7dm9OnTyOEoEmTJlx33XW8\n/PLLhIWFMW3aNHr06MGgQYPIycmhYcOGvPnmm/z5z3/m8OHD3H///Zw4cUKb42n//v2cPn2aJk2a\ncPLkSe644w7GjRvHq6++ysmTJ+nSpQvXXXcdAwcOJDMzk4yMDO68804WLVpEZmYmM2bM4NNPP8Xj\n8VBeXs6QIUOYNm0ad9xxB8eOHcPlcpGVlcW8efO44447SElJobCwkC+//JKwsLC6vk3KVUqNLq4o\nv8Lx48c5ceIEU6dOpW3btsyZMwcpJWfPnuXzzz8nNzeXhg0bkpiYiMViYc2aNXg8Hq20UlRURFxc\nHB06dCArKwufz4fRaMTn8/Hmm2+yYMECXC4XTZs2xePxEBMTQ05ODlarlU8//ZS8vDxiY2PZsGED\nJSUlWK1WWrVqxbFjxxg4cCDLly+nVatWhIaG8uc//5np06czdepUFixYQPfu3dm0aROPPfYYlZWV\nGI1G5s+fzxtvvEFeXh5z587VgtHAgQOZO3cue/fu5cMPP+Sf//wnRUVFbN++nW3btrFs2TIOHz7M\n6tWrad68OQ888ABOp1Oliyu/CzUio6IEKCwsJDU1lT/96U/07NmT0tJSgoKCSEhI4PTp03z++efM\nmTOHDRs20L9/f/r27cuqVatwOp1888032lhymzdvxufzcfbsWV588UVWr15Ns2bN6NevH1JKiouL\nWbJkCbfffjs6nQ6v18trr71G//79SUpKYsaMGezZs4fTp09z9uxZNmzYQEhICJs2beLNN9/EaDSy\nYsUKMjIyKCwsZOXKlTz22GPceeedREZGcuDAAQ4fPoxer2fw4MHs3LmT4OBgdDodmzdvplWrVpSW\nlrJlyxYefvhhHA4HAwcO5IsvvqCwsJCHH36YGTNmMGnSJAwGA+fOncPn87Fs2TKmTZtWx3dJ+W+n\nSkyKEuDIkSOkp6fjcrnYv38/nTt3xuPxsGzZMtq1a8e2bduwWCykpqYydepUXnnlFW699VZtSowd\nO3bQrl07hg0bxmuvvcYtt9zC5s2bsVqtLFu2jKNHj9KgQQPat2/PXXfdxbFjx8jOzmbLli2EhYWx\nfv16SktLSUpKolevXtp1mc1mHA4HJpOJM2fOMGnSJI4fP47X66Vnz57s3buXKVOm8Omnn2K1WklL\nS6O4uBiDwUCHDh3YuXMnLVu2pLi4mLCwMM6cOcOUKVN4+OGH2bRpExkZGcyePZutW7fSq1cv5s+f\nz8SJE3nuuecYMWIEK1euJDw8nPfee68O745yrVAlJkUJsGXLFsLDw5k9ezajR4/W5i8KDQ1Fr9eT\nk5ODwWBg7NixmEwmiouLeeWVV0hKSiI3N5fy8nIqKipIT0/nu+++Y8yYMaxfvx6z2cxtt91Gs2bN\nWLt2LZ9//jmpqamUlZXhdrsxm824XC7Ky8sxmUzMnj2b+vXrY7FYSE5OZvLkyWzfvp2+ffvy9ttv\ns2jRIp5++mm2b9/O+++/zzPPPMPXX39Ny5Yt+cc//kFwcDAej4eioiLS09MxGo3Uq1cPg8HAc889\nR9OmTSktLeWbb75BSsnWrVu17L+pU6eyfft25s2bR1xcHGVlZXTv3l3rvJuZmVnXt0n5L6eSHxQl\ngH8uJX8qdWRkJF27dsVut2M0GsnIyOD2228nPT2dpKQkdu3ahcViQQiBzWYjNTWVEydO0K5dO3bv\n3s1dd93F4sWLSUxMJCsri6SkJOx2O/Hx8Zw6dYqJEyeSkZHBrl27cLvdxMXFYbVa6dixI61atWL+\n/PkYDAZiY2MZMGAATZo04eWXX8bpdDJs2DDOnTuHzWbDbDZjsVgYNWoUL7/8Mh06dKBLly688sor\n1K9fn6effpoxY8ZQv359PB4P119/Pfn5+TRo0IB27dphMpmYPn06b7/9Nunp6WRmZjJs2DCWL19O\nfHw8rVu3ZsiQIcyYMYMdO3bU9W1SrlIq+UFRfoV//OMfWonAbDYzb948Vq9eTU5ODnq9XhtxvLy8\nnEmTJtGrVy8aNWpEaWkpPp9P6/QaFhaGwWDAarXy1FNP4XA4CAkJITw8nE6dOnHs2DGioqKYM2cO\npaWl/OUvfyE4OFirstu5cyebNm0iNDSUlJQUCgoKOHToEJ988gnh4eGEh4dz6tQpjh8/jl6v58iR\nIxQVFTFz5kwmT57Mxo0bycrKomPHjlRWVjJ79mwmT57MPffcQ6NGjWjYsCGjR4+ma9eu7Nmzhzlz\n5nDvvfeyYcMGOnTowJkzZ1i4cCFTp07lqaeeYujQodx999307t27ju+Qci1QJSZFCTBt2jRef/11\nHnroIaZNm4bZbOb8+fNs3bqVwYMHM23aNEwmE+vWrePLL78kJSWFqqoq3G43H3zwAe+88w7du3fn\nn//8J82bN8dsNuPz+cjIyCA+Pp6cnBx69epFVFQUK1eupH79+hw8eJCEhASsVispKSn06NGDI0eO\n4PV6OXToEG63m8TERIqLi9HpdEgpsdvttGjRgvDwcBwOB3l5eXTr1o1Vq1Zx7tw5Hn/8cTZv3ozH\n46FRo0Y4HA7Onj1LcnIy9erV4/jx41itVmJjY6moqKBx48Zs2bKF4cOHc+DAAUpKSujcuTMnTpyg\npKSE77//nhEjRrBx40aio6Pr+jYpVylVYlKUXyEoKIjo6GhWrFhB27ZtadWqFY0aNeLBBx8kOTmZ\n+fPn895775GZmUlpaSnZ2dm0a9eOyMhInn32WfLy8jhx4gQjR47kyJEjHDx4kPT0dKZMmULnzp35\n+OOPKS8vZ8GCBTRu3JiBAwfy0EMPER4eTlhYGKdPn9ay5Vq1aoVOpyMyMpKcnBz69OmD3W4nLi6O\nRx55hJSUFEpKStizZw+xsbEsW7aMoKAgevXqRYMGDYiIiGDatGm4XC5GjRpFQUEBycnJnDhxgoiI\nCMrKyhg5ciTJycnk5eURGhpKRkYGMTExDBo0iKSkJAoLC6moqCAtLY3MzEzOnz9f17dIuQaowKQo\nAXw+HyUlJXTs2JFz587RuXNnHnzwQRwOB/n5+YSFhbFr1y527NiB0WgkLi6Ob7/9VuuL5B+258iR\nI3z//fd0794dg8GAz+fj+eef19psGjVqxKOPPkp4eLg2YZ/RaOTmm2/myJEjtG7dmrS0NMLCwigu\nLsbn87Fy5UrcbjfJycnodDrS0tKw2WwYjUbcbjdCCLxeLy1btkRKSWVlJU899RT79u3jpZdeIiIi\ngoKCAk6ePEnTpk3p0qULn3zyCXv27OHcuXM4nU6ys7PZunUrFouFP/7xj5hMJnQ6HR6Ph/nz59Ox\nY0eqqqrq+jYp/+VUYFKUAHfffTdVVVUsX74cu93OoEGD+OyzzwBo3bo1GRkZtGzZkpCQECIiIigs\nLEQIQZs2bcjMzEQIgZSSjIwMxo0bR0lJCcnJyZjNZr777juWLFnC6tWrCQoK4tNPP+Xbb7/F5/MR\nGRlJQkICq1ev5vDhw3To0IEVK1aQl5eH2+3G4/EwceJEfD4fO3bs4L333qNXr154vV5atWpFcXEx\nQUFBpKamcvToUW644QZuuukm4uLi6NGjBy+88ALDhg0jKCiI0aNHs2jRIkpLS8nLy8Pn89G3b1/m\nz5/P0KFDgeqhme644w4cDgdt2rQhNTWVSZMm0bBhQ0JCQuryFinXAJUurigB6tevT0xMDAUFBURH\nR3Pffffh9XopKSkhMTGRtLQ0LYGhqqqKbdu2ERcXx6FDhwgODqa8vFybYmL9+vVERERgt9v505/+\nRIsWLbQU85tvvpkDBw6QlZWF0Wjk/vvvZ86cORiNRux2O+vXr+f+++9n06ZNGAwGoqOj+eabb9Dr\n9ZSVlTFixAheeOEFOnToQGJiIk2bNmXz5s3Y7Xa++eYbhg8fTnJyMiUlJZw+fZrExETWrl2rpZA/\n9thjfPTRR9hsNm655RbsdjshISGcOnVKq0LMz8+nefPmOBwOIiMjadGiBX369KnrW6RcA1Tyg6IE\n2Lt3Lz169CAyMpLg4GCKi4uxWCyUl5cjpUSv13Pq1Cluv/129u7dS3R0NA6Hg+bNmwNgtVqJiYkh\nNjaWb7/9lqqqKnQ6HR07dmTPnj20aNGCkpISXC6Xlk6+Z88ewsPDtREWOnfuzMmTJxkwYAAbN27k\n7NmzREZG4nA4tJHMH3vsMT744AO+++47EhMTadeuHc2bN2f58uUcOnQIi8VCREQEBoOBoqIijEYj\nKSkpHDx4EK/XS3BwMFDdpmaz2fB4PPTu3ZvKykreeusttm3bxrp169ixYwcVFRWYzWbq1aunDVrr\n319R/hMq+UFRfoUnnngCh8PBuXPnSEpKQgiB1WrF6XTSuHFj7r//fqZNm8bw4cPxer1IKfH5fERF\nRbFnzx6OHz/Onj17yM7OxuFw8Mgjj9C2bVv2799PREQEJ0+epGvXrsTExLBkyRKysrK47bbbCAsL\nw+FwEBoaSlFREfXr12fhwoUUFhaSnJyMzWYjJiaGVq1aUVlZyT333MOePXsYPny4NivtyZMnOXLk\nCGazmaqqKgYPHqwFwfbt21NaWkpCQgLBwcHaxIJ6vZ7mzZvTpUsXDh06xNKlSxk/fjxr1qyhrKyM\n0tJSXC4XQUFBnD9/nltvvVUFJeWyUyUmRQkwfPhwPvvsM0JDQ7VUcYfDQVhYGFJKbDYbXbp04fHH\nH2fx4sVs374dr9dLSEgI8fHx7N+/n/DwcHQ6HWPGjGHx4sXY7XZCQ0NJSkrCZrORm5uLzWYDIDU1\nlfz8fPR6PfXq1SMvLw+DwUBVVRU333wzW7ZswW63k5SUxL333svrr79OSEgIRqORsrIyvF4v5eXl\nAMTExGC327n33nvJysqiZ8+evP/++wDodDoqKyspKCggJCQEvV6Pw+HAYDDgdDrxer0kJibSokUL\nNm/eTHx8PFarldDQUM6fP4/H46FZs2Z06dKFRYsW1dXtUa5yv7TEVCeBSQhxCjgPeAGPlLKTECIa\n+BhIBk4Bf5BSlorqUTFnAYMAGzBeSrnn3x1fBSbl1xoxYgSrVq0iMTGRoqIiOnXqREJCAgcPHtT6\nC9ntdpo2bcqZM2cIDQ2lsrISp9OJxWJBr9dTVVWFlJKUlBSysrIwGAzo9XqcTidt2rQhJydHa9Np\n1qwZ+fn5FBUVYTab8Xq92O12XnzxRf76178ihMDn82EymTAajQghaN++PSdPniQnJ4fw8HAqKiqI\niIjA6XRiMBhwu904nU4ATCYTUkpuuukmtmzZQllZGcHBwbjdbnw+HwA33XQTO3fupKKiAqiu3tPp\ndLhcLqSUdOnShZiYGNatW8egQYNYvnx5nd0f5ep2NVTl3SilvC7gIv8CrJdSNgPW13wGGAg0q3lN\nAN753a9UuWZMmTKFiIgIdDod58+fZ/v27SxbtgyTycQjjzzCDTfcwA033EBWVhYul4uysjJ0Oh0G\ng4HKykrKysqQUrJz505KS0sJCQmhffv22gR8/u379OlDfHw8R48epaCgALPZzJ/+9CctoEyZMgW9\nXo9Op0On02E0GnE4HAQFBbFp0yZycnIwmUza9h6PRzuHP0ABPPvss9x0000cOnSIsrIywsLC8Hq9\n+Hw+QkNDMRgM7Nu3j7i4OMLCwmjSpIn2vRACg8FAWloaX3/9NXFxcXi93jq4K8q1pi5LTJ2klEUB\n6zKBvjVTqjcAvpFSthBCzKlZ/vDC7X7q+KrEpPxab731Fq+++ipnzpzBYrGQlJSEXq+noqKC0tJS\nbYTv/v37s3z5coKDg7FYLJSUVE+sbDKZsFgs3HvvvXz00UcUFBTg8XgAtAe92+0mODiY2NhYrd+U\n0+kkIiKClJQUDh8+jMFgwOVy4fF4tPYnj8eDEIKwsDCcTidSSpxOJ71798btdmM0GtmxYwdCCOx2\nOyaTSSv1+IOnx+PRMv/0er0WaBISEsjPz8dsNiOEwOl0EhUVBUBpaSlNmzaloKCAxMREDhw4UDc3\nR7nqXeklJgl87AmEaQAAESlJREFULYTYLYTwz9VcLyDYnAPq1SwnAjkB++bWrFOUS27jxo0UFhYS\nFRVFUFAQx48f58iRI+Tn52Oz2fD5fHg8HtauXYuUEqPRSFVVFYmJiQgh0Ov1lJSUsGTJEhwOB1FR\nUVonVX+7lb+qzGq1YrFYaiUTZGVl4fV6cblc6PV6hBB4PB5tf4PBQHl5OUFBQej1egB27NjB4cOH\nadasGR6PB7fbDVSXonw+H1JKEhMTqVevHo0bN9b2S0pKIiQkhOjoaM6dO4der8fn8xEfH0+7du0o\nLCzE7XYTERHB0aNHKSsr49Zbb/39b4pyzamrwNRLStmB6mq6R4QQNwR+KauLcf9RUU4IMUEIsUsI\nsauwsPASXqpyLZkyZQotW7ZEp9MRFRWlpYhbLBZtG5PJRGxsLBaLBZfLRXh4OJWVlZjNZnS66j8p\np9NJcXExdrudBg0aIISgqqpKG53Bn/p95swZLRGivLycqqoq6tWrp5VapJRIKfF6vZhMJoKDgxFC\nUFlZid1uByA+Pp4nn3ySJUuWYDAYMBgMDB48mNTUVIQQBAUFYTQayc/PJy8vj8rKSuLj43E4HNhs\nNm1kCp1Op5XM9uzZQ2RkJGVlZRQXF2s/+4wZM37Hu6Fcq+okMEkpz9a8W4HPgS5AQU0VHjXv1prN\nzwINA3ZPqll34THnSik7SSk7xcXFXc7LV/6LCSHIzc3VHsg+nw+fz1drGB5/icPlcjFp0iQKCwsp\nLy/H6XRqwae0tJT27dvTpk0brSRlMBiIj4/XgsDgwYOJj4/Xqvr85/eP9uDn9XrxeDzExcVx0003\nER4ejpSSsLAwhBCYTCbeeOMNpJR0796d1q1b88UXX5Cfn88999yjlbLcbrc2yrnVaqVFixaEhIRQ\nWlqqlc4iIiIYMGAAJpNJGxfPaDRiNBrR6XSYTKbf6U4o17LfPTAJIUKEEGH+ZeB/gIPASmBczWbj\ngBU1yyuBe0S1bkD5v2tfUpTfYvv27TgcDrxer9b+4m8biomJISIiApfLxZAhQ/B4PNo04/6Sir/q\nDCAjI4Pt27dz4sQJXC4XXq+XrKws7Vx79+7FarUSGRlJ69attXNdyO12o9PpOHr0KKtXr9bSz/1V\nfP5EhuTkZDZu3MjJkyeB6sC5detW7HY7gbUI/iB78OBBUlJS0Ol0WCwWIiMjcbvdzJs3T/sdhISE\n4HK5aNasGT6fj3Hjxv3o+hTlUquLIYnqAZ/X/AEagKVSyq+EEDuBT4QQ9wOngT/UbP8l1ani2VSn\ni9/7+1+ycq0ICQnRHtwOhwNAq84rLi7WSkuzZ88G0MbG83g8hIaGallyQUFBtUo9RqMRl8ulbe/1\nehFC4Ha7tVlv/dv5z+vfFtASFTwej5atV1lZCcDx48e1TrNQnawQGRmJ2WzWpsrwp4Y3atRIazsq\nKSmhrKwMn8+H0+nE6XTicrm0YwXuV1BQgJSS66+//jL95hXlB797iUlKeUJK2b7m1UZK+UrN+mIp\n5U1SymZSypullCU166WU8hEpZaqUsq2UUqXbKZdNQUGB9jD2+Xw0a9YMo9GoPfS9Xi9GoxEpJbGx\nsQwcOBCo7sDqbyvS6XQ/Kvn4U7mllAQFBQHVAcW/zt825Q98/vV+gf2OhBBaIPPzDwTrP29ZWRn5\n+fmEhobW2q+0tBS73V4reSI8PLzWNfoDqs/nw263o9PpKCqqTqB9+OGHteMpyuWihiRSlAANG1Y3\nZxoMBoQQ2iywKSkpNGnSBEB7sNvtdtatWwdUP/Svu+46rZThf8jDD8HGPwSQv4rQYDAQHBxMeHi4\n1s7kdDpp27ZtrQDlP76fP4gFXnNiYiKVlZVYLBbCwsIYOnQoRqNRCygNGjTQ+mbpdDrtZ/CXlgL5\nfD4tySIqKkobB9D/3cWqGxXlUlJDEilKgK5du7Jjx45aVWP+vkP+fkWB/ZIC/36Cg4PR6/Vayeli\n/B1cgVolLH9CxE+VRkwmk1bF5xcbG6sFnkAXViP+Fv6+T35ms/nf/nyK8u9c6f2YFOWK5PV60el0\ntUoR/g62/lRqv7i4uFrVdmFhYVrnWT//qN6hoaFAdWnLbrfXerj7S0D+oHRhiQiq27suLKkEpnEH\nslgshIeHYzQaiYqK0kaBgIsnVwSmwgcuQ3UgCtzXn6KuKJeTCkyKEsDfJhP4AA8MRoEPeavVWisw\nlZWV4Xa7CQkJ0dZFRUVRVVWlJSp4PB6tlOUPQIHHh+o2n4SEBBITa/cjv7B2w79/RERErc/l5eVU\nVlbicrkoLS2tdfyL1ZAEltxsNhvR0dEX/U5KqarxlN+FCkyKEmDEiBFA7Qe40WgkLCwMqB1EjEaj\nNrqC/zuDwUBJSUmtNiKr1crFXKzazj8RYF5e3o+q6fyBx//urxL0jy7uZ7FYflGCgn9QWKBWkoR/\neKXAc/i/a9Gixc8eV1F+KxWYFCXAwoULtWWTyaSNWRfYZhPYwfan+ANYQUEB8OPqOaPReNH9AgdJ\nvTApwR9Efiro+Hw+jEZjrWpC/3TrF/L/XP4AfLE2KX/H4EB5eXkXPbeiXEoqMClKgKZNm2rLgW1K\n/sFbASIjI2tV6QXyD7QaFBSkZeFB7WDSvXt3LahdWDV2YSAI5O/75E/vDuQPdP5UdPghGSM3N/ei\n1xnIfz3+kiGgDYcU+LNWVFSodHHlslOBSVECXDitQ2AygD8rrri4+EcP9kBSSm34nwuPJ4Tg2LFj\ntba92Pl/qi1HSklFRcVP9pPyT+XuXxcaGoper//ZWWf91+FPJw8MqIG/A/84gopyOal/YYoSoEGD\nBsAPVW8XS40OLJX8p6SUF82m0+v1WinLv93F+AOGlJLw8PAfBajA6jeLxUJRURE2m02bo+liJS5/\nx9yYmBigOhj5A6R/IkT/zxwYVBXlclGBSVEC+IOGv18R/BCk/CnYF2sfCixF+B/iF253sf2MRqPW\nqddfyrpYZ1p/FZvX69XmSaqoqNA6wvpJKfH5fBgMhh+V1lJSUmqNVO4PguXl5T8ZMP0dcQOv49+1\nrSnKpaACk6IEGDt2LFBdMmnQoIE21YR/qnH/eHf+FO3AwBAVFYXZbNYSCSZPnlwraFysbcblcpGT\nk6N9FxQURHJy8o/6Np0/fx6DwUC9evUoLS3V9nc4HERHRzNixAitbSshIUGbWsN//uDgYG0AWbvd\nrk3jHhcXp6WlB1bZhYaGaqOY+8fRA2jevPlPJm4oyqWiApOiBKhXr3p+SiEEZ8+exWKxkJKSQpcu\nXbSqrpSUFC013D/YKVQPnurz+UhISKBXr15MnToVvV6vlbTCwsK0AViTkpJqndcfQDweD+Xl5Vqp\nKJDP56OgoOBH7UXFxcWkp6cTGRlJVFSUNuAq/FAlGJjh55/g0O12U1RURG5uLtHR0bXazSorK6mo\nqCAsLAyDwYDJZEIIQXx8/G/6/SrKL6GGJFKUAPHx8doUEf9uiKDf6nIe+3IKCgpSVXnKr6aGJFKU\nXyGws+rlDBxXY1CC6nYwFZiUy00FJkUJoB66P0+liyuXW11MFPirCCEGALMAPTBfSjm9ji9J+Z34\nJ8j7T14ulwun04ndbsfhcGjLTqcTh8NRa9nhcGCrGVxV+Xldu3WnefNmmE1mzBYz5poMv597BQcH\na6Np/KcvNUbfteWqCExCCD3wFnALkAvsFEKslFIertsruzr4H+xerxe3231Jl3/Ny+3x4HF7cLvd\n1cseDx6PG4/Hq717vR68Nee6lO2gQmcAvQH0QUidHp8w4BN60BuQul/fP+lakllQiqlRMMLmgkIb\nOulFJz0Inwd8HvB6kB43cOnum06vR6/XYzAY0Ot/HLiCgvzvQQQZflj27/Nrl4OCgmrO+duWVWD9\nz1wVgQnoAmRLKU8ACCE+Am4DrorAdOLECbKzs3/dQ9zt1gKLf9l9wUP9h4Dhwe321DzUvXhqHu6/\nS4KLTo/Q6UDoQOihZlkKHaBDCoEUOiQ160TA97ogEMHVn406CA74zv/SBX7W13wvfljWBWwbcFyp\nN4DOgNRVv/OzD4gPLv/v6iqXMnQSDkvYv99ISpA+8NUELK8H4XMjfN7q9TUvIX0gvdr2P3wvEfIi\n2/oCPvt84PSBwwfSg5CuWtsKKRF4q98DjlN9XdXHlv7jXWZ6vR693oDeoMdgCMKg12MICsJgqAmE\nhiAMNcHVWBPQLsXLYrHQo0ePnxxC60p1tVxtIpAT8DkX6Bq4gRBiAjABoFGjRr/flf0C06ZN0/qQ\nXMmkENUPcWEAnb7mYa7XHupSp6/1WVsWeqgJPAhRHRy44LMQgO6i2/0QZH5YL7lgX/82F65H/IJg\no1xKB99+lA5P/kwA9/+nQadHUp3e/m//eyRl9QvfD0FK/rAM/nVSC1z+9bW2u+i6mn25YN/AgOXz\nIqSn+t13sXdPddD0v8ua91/I6/VWd3iugybMmTNn0qnTzybCXVGuinRxIcQIYICU8oGaz2OBrlLK\nRy+2/ZWWLl5VVUVhYaH2j/NKevmrzTxeb00pq6YardZ2Pm3Z5/Xi9dW8e3/5H+ZlJXQIoQOdP2AF\nBj5/MPQHPKEFPYmoHRD5cYD078MFQfbnA+cVGiz9f++BD2YCHuoy4KFeK0gEPNS5YLuAYCCQNSWV\nnwoctc8lA5brmk6nq64y1OnR6XU1fc70NaWdgJdBj8FftegvBV24zRXyMpvN2jBbV4Jfmi5+tZSY\nzgINAz4n1ay7KoSEhBASElLXl3FZ+MdV+09ev2afyxeYa4JyQJD2er1aoA4Mxv4gHRicr4b/2P1S\n1Q/lwIfzTz/waj2QDYYr9sH8S1+BEz4qde9qCUw7gWZCiBSqA9KdwJi6vSQFav6XqdP9poFNr2Y+\nn++igfZKDlj+0cPVg1m5Ul0VgUlK6RFCPAqsoTpdfIGU8lAdX5aiaIH5amtcVpQr2VXz1ySl/BL4\nsq6vQ1EURbm8VBduRVEU5YqiApOiKIpyRVGBSVEURbmiqMCkKIqiXFFUYFIURVGuKFfFyA//KSFE\nIXC6rq9DUS6RWKCori9CUS6BxlLKuJ/b6L8yMCnKfxMhxK5fMoyLovy3UFV5iqIoyhVFBSZFURTl\niqICk6Jc+ebW9QUoyu9JtTEpiqIoVxRVYlIURVGuKCowKcp/ISFEXyFEj7q+DkX5NVRgUpT/Tn0B\nFZiUq5IKTIpyiQghkoUQR4QQ84QQh4QQXwshzD+xbVMhxDohRIYQYo8QIlVUe00IcVAIcUAIMapm\n275CiNUB+84WQoyvWT4lhHip5hgHhBAthRDJwB+BSUKIfUKI3kKIkTXHzRBCbLrsvwxF+Q2umvmY\nFOUq0QwYLaV8UAjxCTAcWHyR7ZYA06WUnwshTFT/J/EO4DqgPdWjPez8hUGkSErZQQjx/4AnpZQP\nCCHeBSqllDMBhBAHgP5SyrNCiMjf/FMqymWkSkyKcmmdlFLuq1neDSRfuIEQIgxIlFJ+DiCldEgp\nbUAv4EMppVdKWQB8C3T+Bef87N+dr8b3wPtCiAepngVaUa5YKjApyqXlDFj2cmlqJTzU/ls1/cQ5\nf/J8Uso/An8FGgK7hRAxl+C6FOWyUIFJUX5nUsrzQK4QYhiAECJYCGEBvgNGCSH0Qog44AZgB9UD\nEreu2S4SuOkXnOY8EOb/IIRIlVJul1K+ABRSHaAU5Yqk2pgUpW6MBeYIIaYCbmAk8DnQHcgAJPC0\nlPIcQE171UHgJLD3Fxx/FfCpEOI24DGqEyGaAQJYX3MORbkiqZEfFEVRlCuKqspTFEVRriiqKk9R\nLiMhxFtAzwtWz5JSvlcX16MoVwNVlacoiqJcUVRVnqIoinJFUYFJURRFuaKowKQoiqJcUVRgUhRF\nUa4oKjApiqIoVxQVmBRFUZQryv8HQ4ZrKG/0uFcAAAAASUVORK5CYII=\n",
            "text/plain": [
              "<Figure size 538.128x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Bvw7vImuYlw6",
        "colab_type": "code",
        "outputId": "5ed1520b-cebb-4768-f1fd-19b18aad151c",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!ls"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "10xv2_whitelist.txt\t\t\t ClickTagsBig2\n",
            "96S_BIG2_tags_S1_L001_R1_001.fastq.gz\t ClickTagsSmall1\n",
            "96S_BIG2_tags_S1_L001_R2_001.fastq.gz\t ClickTagsSmall2\n",
            "96S_SMALL2_tags_S2_L001_R1_001.fastq.gz  ClickTagWhitelist.csv\n",
            "96S_SMALL2_tags_S2_L001_R2_001.fastq.gz  featuremap.py\n",
            "96S_TagsBig1_S1_L001_R1_001.fastq.gz\t FeaturesMismatch.fa\n",
            "96S_TagsBig1_S1_L001_R2_001.fastq.gz\t FeaturesMismatch.idx\n",
            "96S_tags_S1_L001_R1_001.fastq.gz\t FeaturesMismatch.t2g\n",
            "96S_tags_S1_L001_R2_001.fastq.gz\t kallisto\n",
            "bustools\t\t\t\t kallisto_linux-v0.46.0.tar.gz\n",
            "bustools_linux-v0.39.3.tar.gz\t\t Mus_musculus.GRCm38.cdna.all.idx\n",
            "ClickTagsBig1\t\t\t\t transcripts_to_genes.txt\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_lQ-JNcg7KM0",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "OPTIONAL checkpoint save\n",
        "---\n",
        "Use these codeblocks to save the processed data to Google Drive storage. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xw7Z309YZPYo",
        "colab_type": "code",
        "outputId": "beb557fb-6ce2-4810-97b9-01c1b975d105",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "!mkdir ./drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "mkdir: cannot create directory ‘./drive/My Drive/Colab Datasets/kallisto_20191011/ClickTags’: No such file or directory\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eCuZLXXdYoHI",
        "colab_type": "code",
        "outputId": "a338dbff-1ef8-4d58-c957-756c3d6db90b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 68
        }
      },
      "source": [
        "!rsync -rvzh ./ClickTagsBig1/ ./drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsBig1"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "sending incremental file list\n",
            "rsync: mkdir \"/content/./drive/My Drive/Colab Datasets/kallisto_20191011/ClickTags/ClickTagsBig1\" failed: No such file or directory (2)\n",
            "rsync error: error in file IO (code 11) at main.c(675) [Receiver=3.1.2]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UcgZ4zV2Y0_S",
        "colab_type": "code",
        "outputId": "59ec3db8-d450-4dd7-9a73-139961ad6ab6",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 88
        }
      },
      "source": [
        "!rsync -rvzh ./ClickTagsSmall1/ ./drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsSmall1"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "sending incremental file list\n",
            "rsync: mkdir \"/content/./drive/My Drive/Colab Datasets/kallisto_20191011/ClickTags/ClickTagsSmall1\" failed: No such file or directory (2)\n",
            "rsync error: error in file IO (code 11) at main.c(675) [Receiver=3.1.2]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XjJAe4CgZTpK",
        "colab_type": "code",
        "outputId": "3c5b755f-a492-4cf5-d2ee-ece7213fea8e",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 68
        }
      },
      "source": [
        "!rsync -rvzh ./ClickTagsBig2/ ./drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsBig2"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "sending incremental file list\n",
            "rsync: mkdir \"/content/./drive/My Drive/Colab Datasets/kallisto_20191011/ClickTags/ClickTagsBig2\" failed: No such file or directory (2)\n",
            "rsync error: error in file IO (code 11) at main.c(675) [Receiver=3.1.2]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ExMke0JdZTX7",
        "colab_type": "code",
        "outputId": "b80a615d-24cb-4ad1-d90a-f1a0a337cfc1",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 88
        }
      },
      "source": [
        "!rsync -rvzh ./ClickTagsSmall2/ ./drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsSmall2"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "sending incremental file list\n",
            "rsync: mkdir \"/content/./drive/My Drive/Colab Datasets/kallisto_20191011/ClickTags/ClickTagsSmall2\" failed: No such file or directory (2)\n",
            "rsync error: error in file IO (code 11) at main.c(675) [Receiver=3.1.2]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nM9cGjcZgqw3",
        "colab_type": "code",
        "outputId": "32899022-efff-4a2d-be34-db673cb5fee8",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "!ls sample_data"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "ls: cannot access 'sample_data': No such file or directory\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Kst1Asr27WSh",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Download the cDNA data\n",
        "---"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gBJoMu-6o_cB",
        "colab_type": "code",
        "outputId": "68c26235-f89d-45d7-988b-85ae94366fae",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 68
        }
      },
      "source": [
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/eed2447c-787d-409a-a274-de8e8d3a2fc0/ -o FT-SA08522_S1_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/13d0b71d-9c41-44f5-bd45-2b374aa5b704/ -o FT-SA08522_S1_L001_R2_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/b686466d-b538-4e81-a378-2c4156c6177e/ -o FT-SA08523_S2_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/cdbe5d53-6d3a-482a-b2f5-d8de7f89d910/ -o FT-SA08523_S2_L001_R2_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/eb869388-66a0-4fb1-bf44-cc26a66f3764/ -o FT-SA08524_S3_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/7863fdd4-ac6d-4a8f-919b-84a55c38c8e8/ -o FT-SA08524_S3_L001_R2_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/8615f530-ae0d-45f4-aad5-1e2fc3e4a054/ -o FT-SA08525_S4_L001_R1_001.fastq.gz\n",
        "!curl -L -C - https://data.caltech.edu/tindfiles/serve/35dfe547-ed6c-4dfc-9de8-e8e4d20ad088/ -o FT-SA08525_S4_L001_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            " 86 16.2G   86 14.0G    0     0  18.4M      0  0:15:02  0:13:03  0:01:59 21.3M"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bPFLbf6LrBUT",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!curl -L https://data.caltech.edu/tindfiles/serve/900437b7-b192-4f33-80db-37a1750e1d66/ -o SA08522_S1_L007_R1_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/368145cc-b3fe-4faa-891d-f3771138d2b0/ -o SA08522_S1_L007_R2_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/c5c4b0da-2e0a-4bec-b1e3-08a7fbeec575/ -o SA08523_S2_L007_R1_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/4675e4ba-4674-4ef8-8ebb-71a8d4e481a2/ -o SA08523_S2_L007_R2_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/4d29d99b-ef41-4482-a350-0329d5bdce8b/ -o SA08524_S3_L007_R1_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/1f8aa4d4-e8a1-49dd-ba2b-ccb78601348d/ -o SA08524_S3_L007_R2_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/eb21980a-22bc-4da5-b5ba-f781a79b2758/ -o SA08525_S4_L007_R1_001.fastq.gz\n",
        "!curl -L https://data.caltech.edu/tindfiles/serve/ff991f43-494c-499a-8e91-e66abf81225d/ -o SA08525_S4_L007_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GQnkLF_Kr2Qz",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls -lh"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3B6XaPJrVYzx",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z6ktuO29OsiW",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Use kallisto | bustools to process the cDNA data\n",
        "---"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aCNsTbHCSkoj",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!time kallisto bus -i Mus_musculus.GRCm38.cdna.all.idx -x 10xv2 -t 2 -o bus_out_SA08522 \\\n",
        "FT-SA08522_S1_L001_R1_001.fastq.gz FT-SA08522_S1_L001_R2_001.fastq.gz \\\n",
        "SA08522_S1_L007_R1_001.fastq.gz SA08522_S1_L007_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9ULHCNfxUYoB",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!time kallisto bus -i Mus_musculus.GRCm38.cdna.all.idx -x 10xv2 -t 2 -o bus_out_SA08523 \\\n",
        "FT-SA08523_S2_L001_R1_001.fastq.gz FT-SA08523_S2_L001_R2_001.fastq.gz \\\n",
        "SA08523_S2_L007_R1_001.fastq.gz SA08523_S2_L007_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mgFa-5NWWNo3",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!time kallisto bus -i Mus_musculus.GRCm38.cdna.all.idx -x 10xv2 -t 2 -o bus_out_SA08524 \\\n",
        "FT-SA08524_S3_L001_R1_001.fastq.gz FT-SA08524_S3_L001_R2_001.fastq.gz \\\n",
        "SA08524_S3_L007_R1_001.fastq.gz SA08524_S3_L007_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FC2ZJgelWVaK",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!time kallisto bus -i Mus_musculus.GRCm38.cdna.all.idx -x 10xv2 -t 2 -o bus_out_SA08525 \\\n",
        "FT-SA08525_S4_L001_R1_001.fastq.gz FT-SA08525_S4_L001_R2_001.fastq.gz \\\n",
        "SA08525_S4_L007_R1_001.fastq.gz SA08525_S4_L007_R2_001.fastq.gz"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0RxA0PE3WcMt",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0rKovlYsDfUE",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls bus_out_SA08522/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xpoFF7i2wd8B",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!mkdir bus_out_SA08522/genecount/ bus_out_SA08522/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p bus_out_SA08522/output.bus | bustools sort -T bus_out_SA08522/tmp/ -t 2 -p - | bustools count -o bus_out_SA08522/genecount/genes -g transcripts_to_genes.txt -e bus_out_SA08522/matrix.ec -t bus_out_SA08522/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8uI12YDkESkq",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls -lh bus_out_SA08522"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nLfE8zLjEWTB",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls -lh bus_out_SA08522/genecount"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uBzyuUD5Dx9l",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!mkdir bus_out_SA08523/genecount/ bus_out_SA08523/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p bus_out_SA08523/output.bus | bustools sort -T bus_out_SA08523/tmp/ -t 2 -p - | bustools count -o bus_out_SA08523/genecount/genes -g transcripts_to_genes.txt -e bus_out_SA08523/matrix.ec -t bus_out_SA08523/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-HlMJIx7EDvh",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!mkdir bus_out_SA08524/genecount/ bus_out_SA08524/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p bus_out_SA08524/output.bus | bustools sort -T bus_out_SA08524/tmp/ -t 2 -p - | bustools count -o bus_out_SA08524/genecount/genes -g transcripts_to_genes.txt -e bus_out_SA08524/matrix.ec -t bus_out_SA08524/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HZUGDjhkEM1G",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!mkdir bus_out_SA08525/genecount/ bus_out_SA08525/tmp/\n",
        "!time bustools correct -w 10xv2_whitelist.txt -p bus_out_SA08525/output.bus | bustools sort -T bus_out_SA08525/tmp/ -t 2 -p - | bustools count -o bus_out_SA08525/genecount/genes -g transcripts_to_genes.txt -e bus_out_SA08525/matrix.ec -t bus_out_SA08525/transcripts.txt --genecounts -"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kLh8ELWQP06o",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "OPTIONAL checkpoint save\n",
        "---\n",
        "Use these codeblocks to save the processed data to Google Drive storage. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "46fS1NQVFKVo",
        "colab_type": "code",
        "cellView": "both",
        "colab": {}
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mSYEI0LuHQHR",
        "colab_type": "code",
        "cellView": "both",
        "colab": {}
      },
      "source": [
        "!mkdir drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PZhsB-meHSOd",
        "colab_type": "code",
        "cellView": "both",
        "colab": {}
      },
      "source": [
        "!rsync -rvzh ./bus_out_SA08522/ drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08522/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ysNreMgUH16N",
        "colab_type": "code",
        "cellView": "both",
        "colab": {}
      },
      "source": [
        "!rsync -rvzh ./bus_out_SA08523/ drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08523/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IZrFVPAOIWAB",
        "colab_type": "code",
        "cellView": "both",
        "colab": {}
      },
      "source": [
        "!rsync -rvzh ./bus_out_SA08524/ drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08524/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "er_X73HrIXXN",
        "colab_type": "code",
        "cellView": "both",
        "colab": {}
      },
      "source": [
        "!rsync -rvzh ./bus_out_SA08525/ drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08525/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jJKQXmdTIYoH",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls -lh ./bus_out_SA08522/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xByGO6qBa57U",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Download previously saved gene count matrices (if necessary)\n",
        "---\n",
        "If you saved gene count matrices to Google Drive above, the following codeblocks will download the relevant data."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "q6WzUm08Y0U0",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!date\n",
        "!ls"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dMXCwbn95Qv2",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!rm -r ./sample_data"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7-FG5Zou5QS3",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!wget https://github.com/BUStools/getting_started/releases/download/getting_started/transcripts_to_genes.txt"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UgrMrQurvYHv",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tx3jKfsdvbLp",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls -lh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08522/genecount/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-hXpFL73wfYq",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!mkdir ./kallisto_20191011/\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08522/genecount/ ./kallisto_20191011/SA08522/\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08523/genecount/ ./kallisto_20191011/SA08523/\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08524/genecount/ ./kallisto_20191011/SA08524/\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/bus_out_SA08525/genecount/ ./kallisto_20191011/SA08525/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "S3OlyxypASIj",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls -lh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsBig1/genecount"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "JoVmW7UrA4Xw",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsBig1/genecount ./kallisto_20191011/ClickTagsBig1\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsBig2/genecount ./kallisto_20191011/ClickTagsBig2\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsSmall1/genecount ./kallisto_20191011/ClickTagsSmall1\n",
        "!rsync -rvzh drive/My\\ Drive/Colab\\ Datasets/kallisto_20191011/ClickTags/ClickTagsSmall2/genecount ./kallisto_20191011/ClickTagsSmall2"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HnCTF9XMGEuI",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls kallisto_20191011/SA08522"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "p3JcbIhWIqlH",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!ls kallisto_20191011/ClickTagsBig1/genecount/"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vJau8O8HknfM",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Start here for matrix processing steps\n",
        "---"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KEomtIr1GSSZ",
        "colab_type": "text"
      },
      "source": [
        "At this point, the cDNA reads have been quantified with kallisto | bustools, and the ClickTag reads have been quantified with kallisto | bustools | kITE. The ClickTag reads need to be merged into a single *ClickTags x Cells* matrix, and the cDNA reads need to be normalized and merged into a single *Genes x Cells* matrix."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8XpKJUYXaort",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!pip install leidenalg\n",
        "!pip install scanpy"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab_type": "code",
        "id": "EB4DrBIHnB88",
        "colab": {}
      },
      "source": [
        "import scanpy as sc\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import pandas as pd"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wxG3RVHJlXKA",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Read in the gene count matrices, gene names (or ClickTag IDs), and cell barcodes\n",
        "---"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gs91AjvPCiyG",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory = 'kallisto_20191011/SA08522/'\n",
        "big1cDNAs = sc.read(directory+'genes.mtx')\n",
        "big1cDNAs.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header=None)[0]\n",
        "big1cDNAs.var_names = pd.read_csv(directory+'genes.genes.txt', header=None, delimiter='\\t')[0]\n",
        "big1cDNAs"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "evmzCaSFCszA",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory = 'kallisto_20191011/SA08523/'\n",
        "big2cDNAs = sc.read(directory+'genes.mtx')\n",
        "big2cDNAs.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header=None)[0]\n",
        "big2cDNAs.var_names = pd.read_csv(directory+'genes.genes.txt', header=None, delimiter='\\t')[0]\n",
        "big2cDNAs"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LrGovH78Ci58",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory = 'kallisto_20191011/SA08524/'\n",
        "small1cDNAs = sc.read(directory+'genes.mtx')\n",
        "small1cDNAs.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header=None)[0]\n",
        "small1cDNAs.var_names = pd.read_csv(directory+'genes.genes.txt', header=None, delimiter='\\t')[0]\n",
        "small1cDNAs"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ULHAuSBSCivF",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory = 'kallisto_20191011/SA08525/'\n",
        "small2cDNAs = sc.read(directory+'genes.mtx')\n",
        "small2cDNAs.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header=None)[0]\n",
        "small2cDNAs.var_names = pd.read_csv(directory+'genes.genes.txt', header=None, delimiter='\\t')[0]\n",
        "small2cDNAs"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WacnRl89GJz1",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory='kallisto_20191011/ClickTagsBig1/genecount/'\n",
        "big1tags = sc.read(directory+'genes.mtx')\n",
        "big1tags.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header = None)[0]\n",
        "big1tags.var_names = pd.read_csv(directory+'genes.genes.txt', header=None)[0]\n",
        "big1tags"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "02tVeR2poLOh",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory='kallisto_20191011/ClickTagsBig2/genecount/'\n",
        "big2tags = sc.read(directory+'genes.mtx')\n",
        "big2tags.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header = None)[0]\n",
        "big2tags.var_names = pd.read_csv(directory+'genes.genes.txt', header=None)[0]\n",
        "big2tags"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nS-G4ZivsdNt",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory='kallisto_20191011/ClickTagsSmall1/genecount/'\n",
        "small1tags = sc.read(directory+'genes.mtx')\n",
        "small1tags.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header = None)[0]\n",
        "small1tags.var_names = pd.read_csv(directory+'genes.genes.txt', header=None)[0]\n",
        "small1tags"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dBGTJd-zsZMI",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "directory='kallisto_20191011/ClickTagsSmall2/genecount/'\n",
        "small2tags = sc.read(directory+'genes.mtx')\n",
        "small2tags.obs_names = pd.read_csv(directory+'genes.barcodes.txt', header = None)[0]\n",
        "small2tags.var_names = pd.read_csv(directory+'genes.genes.txt', header=None)[0]\n",
        "small2tags"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "juAe5z_wCErQ",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Filtering and thresholding\n",
        "---\n",
        "This experiment consists of four sets of cells generated from two lanes of 10x. A small batch of cells from both lanes was split off while still emulsified and prepared as a separate library. Let's took a look at the rank-UMI plots for each of the samples, both ClickTags and cDNAs. \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "t55V1DG2k9wG",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Rank-UMI plots for ClickTags\n",
        "---"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Je9Ute4fHg9e",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=big1tags\n",
        "plt.plot(np.log10(range(len(target.X.sum(axis=1).A1))), np.log10(np.flip(np.sort(target.X.sum(axis=1).A1))))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jOr5M7d4Hg1c",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=big2tags\n",
        "plt.plot(np.log10(range(len(target.X.sum(axis=1).A1))), np.log10(np.flip(np.sort(target.X.sum(axis=1).A1))))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LqHSPP06DjJr",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=small1tags\n",
        "plt.plot(np.log10(range(len(target.X.sum(axis=1).A1))), np.log10(np.flip(np.sort(target.X.sum(axis=1).A1))))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Y-DwkHWoHgqD",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=small2tags\n",
        "plt.plot(np.log10(range(len(target.X.sum(axis=1).A1))), np.log10(np.flip(np.sort(target.X.sum(axis=1).A1))))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XltjGAUrlq0S",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "cDNA Cell Filtering and Normalization\n",
        "---\n",
        "We will use a simple cutoff to filter true cells based on cDNA counts. For each set of reads, the numpy gradient function is used to identify the maximal slope in the Rank-UMI plot."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cSQ5bAuCSDP2",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "targetmatrix=np.sort(big1cDNAs.X.sum(axis=1).A1)\n",
        "target=np.sort(big1cDNAs.X.sum(axis=1).A1)[-20000:-5000]\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(target)))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(target)))\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=target[np.argmax(np.gradient(target))]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(targetmatrix>threshold))\n",
        "plt.plot(range(len(target)), target)\n",
        "plt.hlines(threshold, 0, len(target))\n",
        "plt.show()\n",
        "plt.plot(range(len(np.flip(target))),np.gradient(target))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VzDVr9l3Up8r",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "plt.plot(np.log10(range(len(targetmatrix))), np.log10(np.flip(targetmatrix)))\n",
        "plt.hlines(np.log10(threshold), 0, np.log10(len(targetmatrix)))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KEoqNAqygtX3",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "big1filtered=big1cDNAs[(big1cDNAs.X.sum(axis=1)>threshold).A1].copy()\n",
        "big1filtered"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wPNAH2BgF3y9",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "targetmatrix=np.sort(big2cDNAs.X.sum(axis=1).A1)\n",
        "target=np.sort(big2cDNAs.X.sum(axis=1).A1)[-20000:-5000]\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(target)))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(target)))\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=target[np.argmax(np.gradient(target))]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(targetmatrix>threshold))\n",
        "plt.plot(range(len(target)), target)\n",
        "plt.hlines(threshold, 0, len(target))\n",
        "plt.show()\n",
        "plt.plot(range(len(np.flip(target))),np.gradient(target))\n",
        "plt.show()\n",
        "plt.plot(np.log10(range(len(targetmatrix))), np.log10(np.flip(targetmatrix)))\n",
        "plt.hlines(np.log10(threshold), 0, np.log10(len(targetmatrix)))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "vXnXlt53jNU7",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "big2filtered=big2cDNAs[(big2cDNAs.X.sum(axis=1)>threshold).A1].copy()\n",
        "big2filtered"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TZqVCqxEkjPL",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "targetmatrix=np.sort(small1cDNAs.X.sum(axis=1).A1)\n",
        "target=np.sort(small1cDNAs.X.sum(axis=1).A1)[-5000:-500]\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(target)))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(target)))\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=target[np.argmax(np.gradient(target))]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(targetmatrix>threshold))\n",
        "plt.plot(range(len(target)), target)\n",
        "plt.hlines(threshold, 0, len(target))\n",
        "plt.show()\n",
        "plt.plot(range(len(np.flip(target))),np.gradient(target))\n",
        "plt.show()\n",
        "plt.plot(np.log10(range(len(targetmatrix))), np.log10(np.flip(targetmatrix)))\n",
        "plt.hlines(np.log10(threshold), 0, np.log10(len(targetmatrix)))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3pHcZaIPINos",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "small1filtered=small1cDNAs[(small1cDNAs.X.sum(axis=1)>threshold).A1].copy()\n",
        "small1filtered"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7vId35tBIg7S",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "targetmatrix=np.sort(small2cDNAs.X.sum(axis=1).A1)\n",
        "target=np.sort(small2cDNAs.X.sum(axis=1).A1)[-5000:-500]\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(target)))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(target)))\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=target[np.argmax(np.gradient(target))]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(targetmatrix>threshold))\n",
        "plt.plot(range(len(target)), target)\n",
        "plt.hlines(threshold, 0, len(target))\n",
        "plt.show()\n",
        "plt.plot(range(len(np.flip(target[:-1]))),np.diff(target))\n",
        "plt.show()\n",
        "plt.plot(np.log10(range(len(targetmatrix))), np.log10(np.flip(targetmatrix)))\n",
        "plt.hlines(np.log10(750), 0, np.log10(len(targetmatrix)))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wRZIV8E0pr7x",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "small2filtered=small2cDNAs[(small2cDNAs.X.sum(axis=1)>750).A1].copy()\n",
        "small2filtered"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gM70zs7lMwZk",
        "colab_type": "text"
      },
      "source": [
        "The rank-UMI plots from cDNA reads all seem reasonable, with the small samples containing about 1,000 cells and the large samples containing about 10,000 cells. \n",
        "\n",
        "---\n",
        "\n",
        "The filtered cDNA data will be further normalized to 10,000 counts per cell for each individual lane, then the matrices can be concatenated to make a final gene count matrix for all cDNA data from all cells.\n",
        "\n",
        "---\n",
        "\n",
        "Because the ClickTag data are used only for sample identification, they do not require any additional normalization. Concatenate the matrices for each set of ClickTags, then we will proceed with cell filtering and sample assignment. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "h_sWNTBcXcR_",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=big1filtered\n",
        "sc.pp.filter_cells(target, min_counts=1)\n",
        "sc.pp.filter_genes(target, min_counts=1)\n",
        "sc.pp.normalize_per_cell(target, counts_per_cell_after=1e4)\n",
        "big1filtered.obs['n_countslog']=np.log10(target.obs['n_counts'])\n",
        "sc.pp.log1p(target)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4iQAzD1PXc-0",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=big2filtered\n",
        "sc.pp.filter_cells(target, min_counts=1)\n",
        "sc.pp.filter_genes(target, min_counts=1)\n",
        "sc.pp.normalize_per_cell(target, counts_per_cell_after=1e4)\n",
        "big1filtered.obs['n_countslog']=np.log10(target.obs['n_counts'])\n",
        "sc.pp.log1p(target)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "6dJubsopXc4I",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=small1filtered\n",
        "sc.pp.filter_cells(target, min_counts=1)\n",
        "sc.pp.filter_genes(target, min_counts=1)\n",
        "sc.pp.normalize_per_cell(target, counts_per_cell_after=1e4)\n",
        "big1filtered.obs['n_countslog']=np.log10(target.obs['n_counts'])\n",
        "sc.pp.log1p(target)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "oTeGTLyJXcyp",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=small2filtered\n",
        "sc.pp.filter_cells(target, min_counts=1)\n",
        "sc.pp.filter_genes(target, min_counts=1)\n",
        "sc.pp.normalize_per_cell(target, counts_per_cell_after=1e4)\n",
        "big1filtered.obs['n_countslog']=np.log10(target.obs['n_counts'])\n",
        "sc.pp.log1p(target)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "2NF188mDYV-o",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "cDNAdata=sc.AnnData.concatenate(big1filtered, big2filtered, small1filtered, small2filtered)\n",
        "cDNAdata"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ywc_W8uGYjYg",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "cDNAdata.obs_names"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Zg-EcW2TYMfd",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "ClickTag thresholding\n",
        "---"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "97h6_HbjLhAc",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#gather all ClickTag reads using the AnnData concatentate function\n",
        "clicktags = sc.AnnData.concatenate(big1tags, big2tags, small1tags, small2tags)\n",
        "clicktags"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_Vg-bl9aqm65",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "clicktags.obs_names"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bXNfvxfANnFo",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "target=clicktags\n",
        "plt.plot(np.log10(range(len(target.X.sum(axis=1).A1))), np.log10(np.flip(np.sort(target.X.sum(axis=1).A1))))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FblMxq0ohEax",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"\"\"Filter the ClickTag matrix using the cell barcodes from \"\"\"\n",
        "clicktagsfiltered=clicktags[clicktags.obs_names.isin(list(cDNAdata.obs_names))]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7_k8MGIttiWK",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "clicktags"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-FSoMcDPse9F",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "clicktagsfiltered"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "90R3_RLsw4Kl",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.filter_genes(clicktagsfiltered, min_counts=0)\n",
        "sc.pp.filter_cells(clicktagsfiltered, min_counts=0)\n",
        "clicktagsfiltered.obs['n_countslog']=np.log(clicktagsfiltered.obs['n_counts'])\n",
        "logdata=sc.pp.log1p(clicktagsfiltered, copy=True)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "J1ExWX9BygHN",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.normalize_per_cell(logdata, counts_per_cell_after=10000)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "op3hu9yJzJk-",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.regress_out(logdata, ['n_counts'])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Hk6qedaYzJcH",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.pca(logdata, n_comps=20)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f2kHdWOizTd8",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.neighbors(logdata)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "SCgoi9QKzTai",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.umap(logdata)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nnj89-W2zTXh",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.leiden(logdata, resolution=2.8)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gQ8yXEVQzf6C",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(logdata, color=['leiden'], s=20, palette='rainbow')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f0rylvO2mi27",
        "colab_type": "text"
      },
      "source": [
        "Clustering the filtered, normalized, and log-transformed ClickTag data already gives 76 clusters that correspond well with the experimental design. But this procedure fails to cluster out samples with very low numbers of cells. To accurately assign all cells to their sample of origin, we will threshold each ClickTag individually using the numpy gradient function to distinguish between positive and negative cells for each ClickTag. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vfPgbotCTeen",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Threshold each barcode individually\n",
        "---\n",
        "The numpy gradient function will again be used to threshold ClickTags. Some custom data ranges were used to the find the maximal gradient in the desired region of the rank-UMI plot, specified at interval_start and interval_end."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m5xD_t97t3CJ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "barcode_threshold=20*[None]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IsJiDi0Uzf2Z",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=0\n",
        "interval_start=10000\n",
        "interval_end=22200\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FsoPjCwkAeuk",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=1\n",
        "interval_start=10000\n",
        "interval_end=22200\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "zroWg4gyAfMy",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=2\n",
        "interval_start=10000\n",
        "interval_end=22200\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sF6HKg94AfU4",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=3\n",
        "interval_start=10000\n",
        "interval_end=22200\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "OyrT-fm8Afkh",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=4\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WOh5nX3YAfft",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=5\n",
        "interval_start=10000\n",
        "interval_end=22200\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IP7xXPTHxEZX",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=6\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "x9LpHwT3AmuX",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=7\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Op_mYNOEAmmJ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=8\n",
        "interval_start=10000\n",
        "interval_end=22200\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aKfy8ux113gB",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=9\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3RVmNMnVAmdb",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=10\n",
        "interval_start=10000\n",
        "interval_end=22000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wKQ9QHz5-4B1",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=11\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.log10(np.sort(logdata.X[:,ibarcode])))\n",
        "plt.hlines(np.log10(threshold), 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GGcooLXD_dqP",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=12\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.log10(np.sort(logdata.X[:,ibarcode])))\n",
        "plt.hlines(np.log10(threshold), 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WiwCmnclAuRo",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=13\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.log10(np.sort(logdata.X[:,ibarcode])))\n",
        "plt.hlines(np.log10(threshold), 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Wj-KXVowAuNs",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=14\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.log10(np.sort(logdata.X[:,ibarcode])))\n",
        "plt.hlines(np.log10(threshold), 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IZ33yQFhAuFO",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=15\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ryN0M3L9At9o",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=16\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UlSAmsGfAt1I",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=17\n",
        "interval_start=10000\n",
        "interval_end=21000\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3LmlUPqhAttZ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=18\n",
        "interval_start=10000\n",
        "interval_end=22100\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aKj45xceAtlb",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Full-length gradient\"\n",
        "ibarcode=19\n",
        "interval_start=10000\n",
        "interval_end=22100\n",
        "plt.plot(range(len(logdata.X[:])),np.gradient(np.sort(logdata.X[:,ibarcode])[:]))\n",
        "print(\"gradient maximum in interval of interest\")\n",
        "print(max(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end])))\n",
        "print(\"position of gradient maximum\")\n",
        "print(np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start)\n",
        "print(\"threshold value at gradient maximum\")\n",
        "threshold=np.sort(logdata.X[:,ibarcode])[np.argmax(np.gradient(np.sort(logdata.X[:,ibarcode])[interval_start:interval_end]))+interval_start]\n",
        "print(threshold)\n",
        "print(\"number of cells above threshold\")\n",
        "print(sum(logdata.X[:,ibarcode]>threshold))\n",
        "plt.show()\n",
        "plt.plot(range(len(logdata.X[:,ibarcode])), np.sort(logdata.X[:,ibarcode]))\n",
        "plt.hlines(threshold, 0, len(logdata.X[:]))\n",
        "barcode_threshold[ibarcode]=threshold"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TenjKWWhvD_1",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#now that we have our threshold values, we can make a binary threshold table and histogram of positive ClickTag calls per cell\n",
        "dflogdata=pd.DataFrame(logdata.X)\n",
        "for i in range(len(barcode_threshold)):\n",
        "    bc=str(20+i)\n",
        "    dflogdata.loc[:,i]=dflogdata.loc[:,i]>barcode_threshold[i]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uJW5wLk1EJRg",
        "colab_type": "text"
      },
      "source": [
        "Below are the number of cells above threshold for each of the ClickTags used in this experiment. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Y4wNl3z-Aogu",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "dflogdata.sum(axis=0)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "vhIHesC4BH_T",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "labels, counts = np.unique(dflogdata.sum(axis=1), return_counts=True)\n",
        "plt.title(\"Number of ClickTags Above Threshold\")\n",
        "plt.bar(labels, counts, align='center')\n",
        "plt.xticks([])\n",
        "plt.ylim(0,24050)\n",
        "plt.grid(False)\n",
        "print(counts)\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Quej4M5MnNLT",
        "colab_type": "text"
      },
      "source": [
        "The above bar chart is very encouraging. Of 22,312 cells identified by cDNA counts, 20,390 cells (91%) are positive for exactly two ClickTags. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ySvJ9CJ4Hly0",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "dflogdata.index=logdata.obs_names\n",
        "tags2=dflogdata[dflogdata.sum(axis=1)==2]\n",
        "tags2=tags2[tags2.loc[:,:11].sum(axis=1)==1]\n",
        "tagsfilteredsinglecondition=tags2[tags2.loc[:,12:].sum(axis=1)==1]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "V5IrGVbKHlfD",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "tagsfilteredsinglecondition.sum(axis=0)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vKOt-EusKXcC",
        "colab_type": "text"
      },
      "source": [
        "The variabletagsfilteredsinglecondition identifies cells that were assigned exactly two ClickTags in accordance with the experimental design. The high concordance with the experimental design gives us confidence that the ClickTag labeling can be used to successfully demultiplex this large experiment. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GdTXBA7Zomm2",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "cDNA Clustering and Embedding\n",
        "---\n",
        "We are finally ready to start exploring the cDNA data. First, a standard single-cell analysis across all cells, ignoring the experimental perturbation. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "B5KL8agBmgam",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "tagsfilteredsinglecondition.index"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "V8V5Zg-TZp4o",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fcDNAdata=cDNAdata[cDNAdata.obs_names.isin(list(tagsfilteredsinglecondition.index))].copy()\n",
        "fcDNAdata"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "22GrPmyWaMw5",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.highest_expr_genes(fcDNAdata, n_top=20)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "09zAhneubIQK",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.violin(fcDNAdata, keys=['n_counts'])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5rLCxr9YbTWJ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "filter_result=sc.pp.highly_variable_genes(fcDNAdata, min_mean=0.0125, max_mean=5, min_disp=0.50, inplace=False)\n",
        "sc.pl.highly_variable_genes(filter_result)\n",
        "print(\"number of highly variable genes: \" + str(sum(x[0] for x in filter_result)))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CUjHy1Njbn2e",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fcDNAdata = fcDNAdata[:, [x[0] for x in filter_result]]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PlghYHticQxK",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fcDNAdata"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3jdbUJPGcpkW",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.regress_out(fcDNAdata, ['n_counts'], n_jobs=15)\n",
        "sc.pp.scale(fcDNAdata, max_value=10)\n",
        "sc.tl.pca(fcDNAdata, svd_solver='arpack')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tvzdHCE2c2H_",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.pca_variance_ratio(fcDNAdata, log=True)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XFXS8jMJc4sL",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pp.neighbors(fcDNAdata, n_pcs=40)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "i3i8Gm28c75J",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.umap(fcDNAdata)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "11TcK8vhc9jU",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.leiden(fcDNAdata)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9pLsdn3peoJz",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "gene_map=pd.read_csv('./transcripts_to_genes.txt', header=None, delimiter='\\t')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "g95G7vMMfHi9",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "gene_map.loc[:,1].values"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "R6NrxxHPBh45",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "gene_map_dict=pd.Series(gene_map.loc[:,2].values,index=gene_map.loc[:,1]).to_dict()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DzcSoNlTgQDn",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "gene_map_dict['ENSMUSG00000090003.1']"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HTsVHBGpf1ni",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fcDNAdata.var_names=[gene_map_dict[i] for i in list(fcDNAdata.var_names)]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "N21NeAjmCkDm",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fcDNAdata.var_names"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "U1LkrKeYc_B3",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=['leiden', 'Hes5'], use_raw=False, s=15)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RXUKMuLidF2W",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.rank_genes_groups(fcDNAdata, 'leiden', method='wilcoxon')\n",
        "sc.pl.rank_genes_groups(fcDNAdata, n_genes=25, sharey=False)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YFCr0x0WE7bB",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=list(fcDNAdata.uns['rank_genes_groups']['names']['5'][:8]))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TvbT-mvlE-NI",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=['Mt3', 'Cst3', 'Pcsk1n', 'Prdx6'], s=20)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xLvvliEhHkh4",
        "colab_type": "text"
      },
      "source": [
        "With clustering and embedding complete, we can move on to sample assignment based on the ClickTag data. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9fcFkLVzotGd",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Sample Assignment from ClickTag data\n",
        "---\n",
        "The arrays BMPkey, EGFkey, ScripDeckey denote the various concentrations of each of the perturbants used in this experiment. The RAkey array denotes whether a sample was treated with retinoic acid (RAkey=1) or Scriptaid/Decitabine (RAkey=0). These 'key' arrays match the layout of the 96-well plate used for this experiment. \n",
        "\n",
        "For sample assignment of each cell, the concentration of each perturbant is determined by referencing the 'key' array for the same position in the well-plate. The concentrations are stored as a Pandas dataframe called expID with the rows corresponding to the cells and the columns denoting the concentration of each perturbant to which that cell was exposed. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "K6BfbBnSFMW4",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Concentration of BMB-4 from highest (3) to lowest (0)\"\n",
        "BMPkey=pd.DataFrame(np.array(['3','3','3','2','2','2','1','1','1','0','0','0']))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Jm1g2uKVIEps",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Concentration of EGE/bFGF from highest (3) to lowest (0)\"\n",
        "EGFkey=pd.DataFrame(np.array(['3','3','2','2','1','1','0','0']))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uit3MUF5IHNL",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"Concentration of Scriptaid/Decitabine from highest (2) to lowest (0)\"\n",
        "ScripDeckey=pd.DataFrame(np.array(['2','1','0','2','1','0','2','1','0','2','1','0']))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "06R5MxBzIJg1",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"1 if the sample is RA, 0 if Scriptaid/Decitabine\"\n",
        "RAkey=pd.DataFrame(np.array(['0','1','0','1','0','1','0','1']))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ePFiE4LzIx-K",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "BMPkeylist=[None]*len(tagsfilteredsinglecondition.index)\n",
        "for i in range(len(tagsfilteredsinglecondition.index)):\n",
        "    BMPkeylist[i] = int(BMPkey[list(tagsfilteredsinglecondition.iloc[i,:12])].values[0][0])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hyJrpX-pIMFN",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "expID=pd.DataFrame(np.array(BMPkeylist), index=list(tagsfilteredsinglecondition.index), columns=['BMP'])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LsxuRQVoI9P4",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "EGFkeylist=[None]*len(tagsfilteredsinglecondition.index)\n",
        "for i in range(len(tagsfilteredsinglecondition.index)):\n",
        "    EGFkeylist[i]=int(EGFkey[list(tagsfilteredsinglecondition.iloc[i,12:])].values[0][0])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7wWMPXOFJAFc",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "expID['EGF']=np.array(EGFkeylist)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "V-5dyTogJBds",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "ScripDeckeylist=[None]*len(tagsfilteredsinglecondition.index)\n",
        "for i in range(len(tagsfilteredsinglecondition.index)):\n",
        "    ScripDeckeylist[i]=int(ScripDeckey[list(tagsfilteredsinglecondition.iloc[i,:12])].values[0][0])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lbM18v3jJC-S",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "expID['ScripDec']=np.array(ScripDeckeylist)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sK4BVLSfJESK",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "RAkeylist=[None]*len(tagsfilteredsinglecondition.index)\n",
        "for i in range(len(tagsfilteredsinglecondition.index)):\n",
        "    RAkeylist[i]=int(RAkey[list(tagsfilteredsinglecondition.iloc[i,12:])].values[0][0])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LXMBVc40JF4D",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "expID['RA']=np.array(RAkeylist)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F4mEmADGJLGB",
        "colab_type": "text"
      },
      "source": [
        "Good, now add a column for plate position and make a heatmap of the experiment."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0XJ_UYfnJHSX",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "expID['PlatePos']=12*(2*(expID['EGF']*-1+3)+expID['RA'])+3*(expID['BMP']*-1+3)+(expID['ScripDec']*-1+2)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5ICHtVBrJJuS",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "temp=expID[expID['BMP'].isin(['1'])]\n",
        "temp[temp['ScripDec'].isin(['0'])].head()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ksd_kz8lJSAR",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "PlatePosArray = np.zeros((8,12))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TtZQ3NKbJhTM",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "for cell in expID['PlatePos']:\n",
        "    PlatePosArray[cell//12][cell%12]+=1"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8bQHuwpWJjQv",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import seaborn as sns"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7auUT-OeJtzE",
        "colab_type": "text"
      },
      "source": [
        "Below is a heatmap representation of the 96-well plate used to culture cells in this experiment. Each position in the heatmap corresponds to the number of cells recovered from that position in the plate. Some conditions induced cell proliferation, resulting in an imbalanced number of cells recovered from each condition. This effect is obvious in the heatmap.  "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MkbCiZJzJmj_",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fig, ax = plt.subplots(figsize=(12,8))\n",
        "sns.heatmap(ax=ax, data=np.log1p(PlatePosArray), annot=PlatePosArray.astype(int), fmt='g',cbar=None)\n",
        "plt.xticks=([])\n",
        "plt.yticks=([])\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eWMoTW0ZpDhZ",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Embedding colored by treatment\n",
        "---\n",
        "Here are UMAP embeddings of the cDNA data colored according to the treatments received by each cell. The trends observed are a clear sign that the cells are responding transcriptionally to the experimental treatment. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9bc4BbREJnIi",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "plt.figure(figsize=(10,8))\n",
        "plt.scatter(fcDNAdata.obsm['X_umap'][:,0], fcDNAdata.obsm['X_umap'][:,1], s=8, c=expID['BMP'])\n",
        "plt.title('BMP-4 Concentration')\n",
        "plt.axis('off')\n",
        "plt.grid(False)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LATqiB_AJ7GI",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "plt.figure(figsize=(10,8))\n",
        "plt.scatter(fcDNAdata.obsm['X_umap'][:,0], fcDNAdata.obsm['X_umap'][:,1], s=8, c=expID['EGF'])\n",
        "plt.title('EGF/bFGF')\n",
        "plt.axis('off')\n",
        "plt.grid(False)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hq6k6kDRK6Fh",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "plt.figure(figsize=(10,8))\n",
        "plt.scatter(fcDNAdata.obsm['X_umap'][:,0], fcDNAdata.obsm['X_umap'][:,1], s=10, c='#C9C9C9')\n",
        "temp=fcDNAdata[expID['RA'].isin([0])]\n",
        "plt.scatter(temp.obsm['X_umap'][:,0], temp.obsm['X_umap'][:,1], s=8, c=expID[expID['RA'].isin([0])]['ScripDec'])\n",
        "plt.title('Scriptaid/Decitabine Concentration')\n",
        "plt.axis('off')\n",
        "plt.grid(False)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZhgDZ7_9K9DA",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "plt.figure(figsize=(10,8))\n",
        "plt.scatter(fcDNAdata.obsm['X_umap'][:,0], fcDNAdata.obsm['X_umap'][:,1], s=10, c='#C9C9C9')\n",
        "temp=fcDNAdata[expID['RA'].isin([1])]\n",
        "plt.scatter(temp.obsm['X_umap'][:,0], temp.obsm['X_umap'][:,1], s=8, c=expID[expID['RA'].isin([1])]['ScripDec'])\n",
        "plt.title('Retinoic Acid Concentration')\n",
        "plt.axis('off')\n",
        "plt.grid(False)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mo2n-nS-qHK_",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Clustering\n",
        "---\n",
        "Here I'm using a little higher resolution for clustering to better match the cell populations generated by the experimental treatments. The cluster assignments are then added as a new column to the expID dataframe. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QwExlUvHLnEb",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.tl.leiden(fcDNAdata, resolution=2.2)\n",
        "sc.pl.umap(fcDNAdata, color=['leiden'], s=20)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "709WFQ2aMIWS",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "expID['leiden']=np.array(fcDNAdata.obs['leiden'])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "91i9WUFsqYeC",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Cluster abundance versus sample assignment\n",
        "---\n",
        "Clusters on x-axis, samples on y-axis. Structure in the heatmap is reflective of the perturbations performed on each cell population."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UbUpf0gqL6Nr",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "num_clusters=len(set(fcDNAdata.obs['leiden']))\n",
        "ClustOccArray = np.zeros((num_clusters,96))\n",
        "for i in range(len(expID.index)):\n",
        "    ClustOccArray[int(expID['leiden'].iloc[i])][int(expID['PlatePos'].iloc[i])]+=1"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "3Rgu8e23MiG7",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "fig, ax = plt.subplots(figsize=(24,12))\n",
        "sns.heatmap(np.log1p(ClustOccArray))\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uHqHqCbzMjx_",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "FlatPlatePosArray=PlatePosArray.flatten()\n",
        "RelClustOccArray=np.zeros((num_clusters,96)).T\n",
        "for i in range(len(FlatPlatePosArray)):\n",
        "    if FlatPlatePosArray[i]:\n",
        "        RelClustOccArray[i]=ClustOccArray.T[i]/FlatPlatePosArray[i]\n",
        "RelClustOccArray=RelClustOccArray.T"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RGBiQpYyMpNT",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fig, ax = plt.subplots(figsize=(24,12))\n",
        "sns.heatmap(RelClustOccArray, cmap='viridis')\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aojy1NWlMrBv",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sorting_list=list(range(12,24))+list(range(36,48))+list(range(60,72))+list(range(84,96))+list(range(0,12))+list(range(24,36))+list(range(48,60))+list(range(72,84))\n"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-3wYg9BhM6F2",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fig, ax = plt.subplots(figsize=(24,8))\n",
        "sns.heatmap(np.log1p(ClustOccArray[:,np.array(sorting_list)]), cbar=False)\n",
        "plt.xticks=[]\n",
        "plt.yticks=[]\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xC-vZtQhM-E5",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "fig, ax = plt.subplots(figsize=(24,8))\n",
        "sns.heatmap(RelClustOccArray[:,np.array(sorting_list)], cmap='viridis',cbar=False)\n",
        "plt.xticks=[]\n",
        "plt.yticks=[]\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FrU8ShVAq0xm",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "PCA of cluster abundance matrix\n",
        "---\n",
        "One way to view the global structure of the cellular response to this perturbation array is to simply perform PCA on the cluster abundance matrix. because the perturbation was 4-dimensional, a 3D PCA can capture most of the variance introduced by the perturbations. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nGwcgh7ZNCYi",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "from sklearn.decomposition import PCA\n",
        "pca=PCA(n_components=3)\n",
        "principalComponents=pca.fit_transform(RelClustOccArray.T)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FL1ccX1ZNqVT",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "from mpl_toolkits import mplot3d"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cBWuS7dTNJlb",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "fig=plt.figure(figsize=(12,12))\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "X, Y, Z = principalComponents[:,0], principalComponents[:,2], principalComponents[:,1]\n",
        "ax.view_init(elev=20, azim=42)\n",
        "ax.scatter(X, Y, Z, c=8*[3,3,3,2,2,2,1,1,1,0,0,0], marker='o', s=200, alpha=.8,cmap='viridis')\n",
        "ax.w_xaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_yaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_zaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_xaxis.set_ticklabels([])\n",
        "ax.w_yaxis.set_ticklabels([])\n",
        "ax.w_zaxis.set_ticklabels([])\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1TNEJLcdNTbm",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "fig=plt.figure(figsize=(12,12))\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "X, Y, Z = principalComponents[:,0], principalComponents[:,2], principalComponents[:,1]\n",
        "ax.view_init(elev=20, azim=42)\n",
        "ax.scatter(X, Y, Z, c=24*[3]+24*[2]+24*[1]+24*[0], marker='o', s=200, alpha=.8,cmap='viridis')\n",
        "ax.w_xaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_yaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_zaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_xaxis.set_ticklabels([])\n",
        "ax.w_yaxis.set_ticklabels([])\n",
        "ax.w_zaxis.set_ticklabels([])\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wDkBY3MvN3lc",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "fig=plt.figure(figsize=(12,12))\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "X, Y, Z = principalComponents[:,0], principalComponents[:,2], principalComponents[:,1]\n",
        "ax.view_init(elev=20, azim=42)\n",
        "ax.scatter(X, Y, Z, c=4*(4*[2,1,0]+12*[0]), marker='o', s=200, alpha=.8,cmap='viridis')\n",
        "ax.w_xaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_yaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_zaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_xaxis.set_ticklabels([])\n",
        "ax.w_yaxis.set_ticklabels([])\n",
        "ax.w_zaxis.set_ticklabels([])\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ye8RORlMN76S",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "%matplotlib inline\n",
        "fig=plt.figure(figsize=(12,12))\n",
        "ax = fig.add_subplot(111, projection='3d')\n",
        "X, Y, Z = principalComponents[:,0], principalComponents[:,2], principalComponents[:,1]\n",
        "ax.view_init(elev=20, azim=42)\n",
        "ax.scatter(X, Y, Z, c=4*(12*[0]+4*[2,1,0]), marker='o', s=200, alpha=.8,cmap='viridis')\n",
        "ax.w_xaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_yaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_zaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))\n",
        "ax.w_xaxis.set_ticklabels([])\n",
        "ax.w_yaxis.set_ticklabels([])\n",
        "ax.w_zaxis.set_ticklabels([])\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sl8hk5d-rPu7",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "Linear Regression Analysis\n",
        "---\n",
        "We'll end with a simple linear regression analysis used to identify genes that are associated with individual perturbations."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-KdTKxcFN_-9",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "\"\"\"Linear Regression Analysis\"\"\""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7sgkWka7ZZ4d",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "BMPdict={0:0,1:0.8,2:4,3:20}"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sKcrP_alZdNs",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "RAdict={0:0,1:1,2:5}"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "30A08Nq5Zn8E",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import copy"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "2f_YUHPUZe8E",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "linreg = copy.deepcopy(expID)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "45cRw9JxZhM9",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "linreg.head()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "q-j0_duYZih_",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import statsmodels.formula.api as smf"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "O0x7Br5HZp7X",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "linreg['_0']=fcDNAdata.X[:,0]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DkSH5XzFZwl2",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "linreg['BMP']=[BMPdict[x] for x in expID['BMP']]\n",
        "linreg['EGF']=[BMPdict[x] for x in expID['EGF']]\n",
        "linreg['ScripDec']=[RAdict[x] for x in expID['ScripDec']*-1*(expID['RA']-1)]\n",
        "linreg['RA']=[RAdict[x] for x in expID['ScripDec']*expID['RA']]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PU3ugc0JZw4-",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "linreg.head()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "alsS7aKLZ3IA",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "all_genes_est_cond = []\n",
        "for counter in range(len(list(fcDNAdata.var_names))):\n",
        "    linreg.iloc[:,6] = fcDNAdata.X[:,counter]\n",
        "    linreg.columns = ['BMP', 'EGF', 'ScripDec', 'RA', 'PlatePos', 'louvain',('_'+str(counter))]\n",
        "    all_genes_est_cond += [smf.ols(formula=(linreg.columns[6]+' ~ BMP + EGF + ScripDec + RA'), data=linreg).fit()]\n",
        "    counter+=1\n",
        "    if counter%100==0:print('counter: ' + str(counter))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gFbqDq9IZ9Om",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "len(all_genes_est_cond)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "36ALCRGPZ_Me",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "all_genes_est_cond[60].summary()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xTVIFxbDaCa1",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "all_genes_est_cond[60].params"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dQzg4C8KaYYc",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "all_genes_est_cond[60].pvalues"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "qBGN_GnwaZeS",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import math"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bPzWZohwadjw",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "BMPpvals=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    BMPpvals+=[all_genes_est_cond[i].pvalues[1]]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lqyyIu42agUk",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "BMPbetas=[]\n",
        "logBMPbetas=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    BMPbetas+=[all_genes_est_cond[i].params[1]]\n",
        "    if BMPbetas[i]<0:\n",
        "        logBMPbetas+=[-math.log(-BMPbetas[i], 2)]\n",
        "    else: logBMPbetas+=[math.log(BMPbetas[i], 2)]\n",
        "        "
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sZYZplC3ahuZ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "EGFbetas=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    EGFbetas+=[all_genes_est_cond[i].params[2]]\n",
        "ScripDecbetas=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    ScripDecbetas+=[all_genes_est_cond[i].params[3]]\n",
        "RAbetas=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    RAbetas+=[all_genes_est_cond[i].params[4]]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "r7K3TOn-ajpk",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "EGFpvals=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    EGFpvals+=[all_genes_est_cond[i].pvalues[2]]\n",
        "ScripDecpvals=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    ScripDecpvals+=[all_genes_est_cond[i].pvalues[3]]\n",
        "RApvals=[]\n",
        "for i in range(len(all_genes_est_cond)):\n",
        "    RApvals+=[all_genes_est_cond[i].pvalues[4]]"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ypkzZcZjalHf",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "testcolors=[]\n",
        "for i in BMPpvals:\n",
        "    if i > math.pow(10, -50):\n",
        "        testcolors+=['#777C7A']\n",
        "    else: testcolors+=['#6DC670']"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-JnTPVb-aqgt",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "plt.scatter(range(len(BMPpvals)), -(np.log10([x+(10**-305) for x in BMPpvals])), s=12, c=testcolors)\n",
        "plt.grid(False)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HNobWjr5avOt",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "plt.scatter(BMPbetas, -(np.log10([x+(10**-305) for x in BMPpvals])), alpha=0.4)\n",
        "plt.show()"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sYSN7ZP3bSd1",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "count=0\n",
        "BMPgenes=[]\n",
        "for i in range(len(BMPpvals)):\n",
        "    if BMPpvals[i] < math.pow(10, -265):\n",
        "        #print(i)\n",
        "        #print(fcDNAdata.var_names[i])\n",
        "        BMPgenes+=[fcDNAdata.var_names[i]]\n",
        "        #print('pval: ' + str(BMPpvals[i]))\n",
        "        #print('beta:' + str(BMPbetas[i]))\n",
        "        count+=1     "
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "c9u6mr-CbkQl",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=BMPgenes, s=20)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Dz_Hv8pebsJp",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "count=0\n",
        "EGFgenes=[]\n",
        "for i in range(len(EGFpvals)):\n",
        "    if EGFpvals[i] < math.pow(10, -100):\n",
        "        #print(i)\n",
        "        #print(fcDNAdata.var_names[i])\n",
        "        #print(EGFpvals[i])\n",
        "        EGFgenes+=[fcDNAdata.var_names[i]]\n",
        "        count+=1       "
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "l1xIgc9dcJ-6",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=EGFgenes, s=20)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wDYFdJnUcM6x",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "count=0\n",
        "ScripDecgenes=[]\n",
        "ScripDecgenesneg=[]\n",
        "for i in range(len(ScripDecpvals)):\n",
        "    if ScripDecpvals[i] < math.pow(10, -50):\n",
        "        #print(i)\n",
        "        #print(fcDNAdata.var_names[i])\n",
        "        #print('pval:' + str(ScripDecpvals[i]))\n",
        "        #print('beta:' + str(ScripDecbetas[i]))\n",
        "        ScripDecgenes+=[fcDNAdata.var_names[i]]\n",
        "        if ScripDecbetas[i]<0:\n",
        "            ScripDecgenesneg+=[fcDNAdata.var_names[i]]\n",
        "        count+=1"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1c0qwFN-cWfS",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=ScripDecgenesneg,s=15)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-sNoXiskca_M",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "count=0\n",
        "RAgenes=[]\n",
        "for i in range(len(RApvals)):\n",
        "    if RApvals[i] < math.pow(10, -250):\n",
        "        #print(i)\n",
        "        #print(fcDNAdata.var_names[i])\n",
        "        #print('pval:' + str(RApvals[i]))\n",
        "        #print('beta:' + str(RAbetas[i]))\n",
        "        RAgenes+=[fcDNAdata.var_names[i]]\n",
        "        count+=1        "
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nYqb8iztcr_Z",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "sc.pl.umap(fcDNAdata, color=RAgenes, s=15)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "msExjIbkcsiU",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "!date"
      ],
      "execution_count": 0,
      "outputs": []
    }
  ]
}